---
title: "MDD integrated analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=TRUE, cache.lazy = FALSE, message=FALSE, warning = FALSE, fig.height=10, fig.width=14)

suppressMessages(library(tidyverse))
library(ComplexHeatmap)
library(circlize)
library(umap)
library(cluster)
library(RColorBrewer)
library(writexl)
library(clue)
#library(UpSetR)

source("../R/MDD_functions.R")
create_pdfs <- TRUE
write_csvs <- TRUE

```

```{r functions}

create_consensus_clustering <- function(mat = multiomics_data_dm,
                                        nbr_clusterings = 5,
                                        nbr_consensus = 3,
                                        clust_num = multiomics_clust_num){
  cons_list <- lapply(seq_len(nbr_consensus), function(c) {
    cons <- cl_ensemble(list = lapply(seq_len(nbr_clusterings), function(b) { #Create clusterings
      medoids <- sample(nrow(mat), size = clust_num, replace = FALSE)
      clusters  <- pam(x = mat, 
                       k = clust_num, 
                       medoids = medoids, 
                       pamonce = 5)})) %>%
      cl_consensus(method = "HE", control = list(nruns = 5))
  }) %>%
    cl_consensus(method = "HE", control = list(nruns = 5))
}

```

```{r setup_parameters_and_metadata}
data_path <- "../multiomics_analysis/data/"

multiomics_all_features_matrix_path <- paste0(data_path, "MDD_multiomics_matrix.csv")
shared_features_path <- paste0(data_path,"MDD_multiomics_shared_features.csv") 
unique_features_path <- paste0(data_path,"MDD_multiomics_unique_features.csv") 
multiomics_clust_num <- 18
cluster_method <- "pam"

experimentalTimePoints <- c("_24", "_48")
ligand_order <-  c("HGF", "OSM", "EGF","BMP2+EGF", "IFNG+EGF", "TGFB+EGF")

condition_order <- paste0(rep(ligand_order, each = length(experimentalTimePoints)), rep(c("_24", "_48"), times = length(ligand_order)))

#create a named vector of color values for the Heatmap clusters
cluster_cols <- c(structure(RColorBrewer::brewer.pal(12, "Paired")[c(1:12)], names = 1:12), c("13" = "#222222", "14" = "#555555", "15" = "#999999", "16" = "#AAAAAA", "17" = "#BBBBBB", "18" = "#CCCCCC", "19" = "#DDDDDD"))

ligand_cols <- c("CTRL" = "#7A4A2A",
                 "PBS" = "#8dd3c7",
                 "HGF" = "#80b1d3",
                 "OSM" = "#fdb462",
                 "EGF" = "#fb8072",
                 "BMP2+EGF" = "#b3de69",
                 "IFNG+EGF" = "#bebada",
                 "TGFB+EGF" = "#ffd92f")

```

The cluster method is `r cluster_method`.  
Input files are `r multiomics_all_features_matrix_path`, `r shared_features_path` and `r unique_features_path`.  

Method summary:

Use replicates to perform differential (DE) analysis between conditions and ctrl_0 for RNAseq, RPPA, GCP and cycIF assays. Use family z scores for motifs.  
rrscale the DE values within each assay  
Use set analysis to select features that are significantly expressed by a single ligand (ligand-specific) or multiple ligands (shared)  
Perform Gap analysis with pam clustering on the shared and ligand-specific values as a single dataset  
Combine 4 similar cluster/module pairs to yield a multiomics_14 dataset  Evaluate the multiomics_14 dataset with UMAP  
Correlate the multiomics_14 modules with IF and cell tracking phenotypes  
Get enriched Reactome pathways for the RNAseq genes each multiomics_14 module  
Evaluate the modules/assay content and distributions  
Evaluate T0 centered versions of the assay values using the multiomics_14 structure  

```{r prepData}

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(multiomics_all_features_matrix_path)) stop("cannot read ",multiomics_all_features_matrix_path)
all_features_data <- read_csv(multiomics_all_features_matrix_path) %>%
  pivot_longer(cols = matches("_24|_48"), names_to = "condition") %>%
  mutate(condition = str_replace(condition, "BMP2","BMP2+EGF"),
         condition = str_replace(condition, "IFNG","IFNG+EGF"),
         condition = str_replace(condition, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_")) %>%
  pivot_wider(names_from = condition, values_from = value) %>%
  select(feature, Type, feature_type, all_of(condition_order)) %>%
  drop_na

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(unique_features_path)) stop("cannot read ",unique_features_path)
set.seed(42) #randomize row positions
unique_features <- read_csv(unique_features_path) %>%
  slice_sample(prop = 1)
unique_features_data <- unique_features%>%
  mutate(ligand = str_replace(ligand, "BMP2","BMP2+EGF"),
         ligand = str_replace(ligand, "IFNG","IFNG+EGF"),
         ligand = str_replace(ligand, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_"),
         Set = ligand) %>%
  select(ligand, feature_type, Abs_Direction, Set) %>%
  rename(Direction = Abs_Direction) %>%
  left_join(all_features_data, by = "feature_type") %>%
  rename(Cluster = ligand) %>%
  mutate(Cluster = paste(Cluster, Direction, sep = "_")) %>%
  drop_na

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(shared_features_path)) stop("cannot read ",shared_features_path)
set.seed(42) #randomize row positions
shared_features <- read_csv(shared_features_path) %>%
  slice_sample(prop = 1)
shared_features_data <- shared_features %>%
  mutate(ligand = str_replace(ligand, "BMP2","BMP2+EGF"),
         ligand = str_replace(ligand, "IFNG","IFNG+EGF"),
         ligand = str_replace(ligand, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_"),
         Set = "shared") %>%
  select(feature_type, Set) %>%
  distinct() %>%
  left_join(all_features_data, by = c("feature_type" ))  %>%
  drop_na

set.seed(42) #randomize row positions
multiomics_data <- bind_rows(shared_features_data, select(unique_features_data, -Direction, -Cluster)) %>%
  slice_sample(prop = 1)
assay_cols <- structure(RColorBrewer::brewer.pal(8, "Paired")[c(3, 2,6,5, 1)], names = unique(multiomics_data$Type))

```
## {.tabset .tabset-fade}

### Combined heatmaps

```{r cluster_data}

#create a numeric matrix with feature row names
multiomics_data_dm <- multiomics_data %>%
  select(all_of(condition_order)) %>%
  as.matrix() 
rownames(multiomics_data_dm) <- multiomics_data$feature_type
multiomics_data_dm <- multiomics_data_dm[!is.nan(multiomics_data_dm[,1]),]

if(file.exists(paste0(data_path,"MDD_multiomics_clusters.csv"))){
  cluster_df <- read_csv(paste0(data_path,"MDD_multiomics_clusters.csv")) 
  clusters <- cluster_df %>%
    pull(Cluster)
  names(clusters) <- cluster_df %>%
    pull(feature_type)
} else {
  #     Consensus clusterings “synthesize” the information in the elements of a cluster ensemble into a single clustering, often by minimizing a criterion function measuring how dissimilar consensus candidates are from the (elements of) the ensemble (the so-called “optimization approach” to consensus clustering).
  # 
  # The most popular criterion functions are of the form L(x) = ∑ w_b d(x_b, x)^p, where d is a suitable dissimilarity measure (see cl_dissimilarity), w_b is the case weight given to element x_b of the ensemble, and p ≥ 1. If p = 1 and minimization is over all possible base clusterings, a consensus solution is called a median of the ensemble; if minimization is restricted to the elements of the ensemble, a consensus solution is called a medoid (see cl_medoid). For p = 2, we obtain least squares consensus partitions and hierarchies (generalized means). See also Gordon (1999) for more information.
  # 
  # If all elements of the ensemble are partitions, the built-in consensus methods compute consensus partitions by minimizing a criterion of the form L(x) = ∑ w_b d(x_b, x)^p over all hard or soft partitions x with a given (maximal) number k of classes. Available built-in methods are as follows.
  nbr_clusterings <- 100
  nbr_consensus <- 3
  nbr_runs <- 25

  set.seed(42)
  res <- lapply(1:nbr_runs, function(x, nbr_clusterings = nbr_clusterings, nbr_consensus = nbr_consensus){
    cons <- create_consensus_clustering(nbr_clusterings = nbr_clusterings, nbr_consensus= nbr_consensus) 
    save(cons, file =paste0("../Data/MDD_multiomics_clstrng_run_",x,".rda"))
  })
  con_cons <- clue::cl_ensemble(list = lapply(1:25, function(x){
    load(paste0("../Data/MDD_multiomics_clstrng_run_",x,".rda"))
    clue::as.cl_partition(cons)
  }))
 
  #generate a consensus of all clustering consensuses
  set.seed(100)
  ce_cons <- clue::cl_consensus(con_cons, method = "HE", control = list(nruns = 5))
  #Save the cluster assignments to disk
  clusters <- cl_class_ids(ce_cons)
  cluster_df <- tibble(Cluster = clusters,
                       feature_type = names(clusters))
  write_csv(cluster_df, paste0(data_path,"MDD_multiomics_clusters.csv")) 
  
  #Calculate the agreement between ensemble elements and the consensus clustering
  #
  #clue::cl_agreement()
  #"FM"  the index of Fowlkes and Mallows (1983), i.e., the ratio N_xy / sqrt(N_x N_y) of the number N_xy of distinct pairs of objects in the same class in both partitions and the geometric mean of the numbers N_x and N_y of distinct pairs of objects in the same class in partition x and partition y, respectively. For soft partitions, (currently) the Fowlkes-Mallows index of the corresponding nearest hard partitions is used.
  
   #calculate agreement between all possible cluster pairings
  agrmnts <- lapply(1:(nbr_runs-1), function(x){
    res <- lapply((x+1):nbr_runs, function(y){
      cl_agreement(con_cons[[x]], con_cons[[y]], method = "FM")
    }) %>%
      unlist()
  })
  
  p_inner_agreements <- ggplot(tibble(agreements = unlist(agrmnts)), aes(x = agreements)) +
    geom_density() +
    theme_bw()
  p_inner_agreements
  
  ags_con <- sapply(con_cons, clue::cl_agreement, y = ce_cons, method = "FM")
    p_agreement_to_consensus <- ggplot(tibble(agreement_to_consensus = ags_con, method = "FM"), aes(x = agreement_to_consensus)) +
    geom_density() +
    theme_bw()
  p_agreement_to_consensus
  
  #Todo annotate figure with statistics
  agreement_to_consensus_mean <- mean(ags_con)
  agreement_to_consensus_sd <- sd(ags_con)

}

#Add the cluster assignments to the data
multiomics_data_ann <-  tibble(feature_type = names(clusters),
                               Cluster = as.character(clusters)) %>%
  right_join(multiomics_data, by = "feature_type") %>%
  mutate(Cluster = factor(Cluster, ordered = TRUE))

multiomics_ann <- multiomics_data_ann %>%
  select(Cluster, Type, Set) %>%
  as.data.frame()

#Create a matrix and heatmap of the mean multiomics values
multiomics_data_mean <- multiomics_data_ann %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

multiomics_count <- multiomics_data_ann %>%
  dplyr::count(Cluster) 

multiomics_data_mean_dm <- multiomics_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_data_mean_dm) <- paste0("Module_",multiomics_data_mean$Cluster,"_", multiomics_count$n)

```

```{r correlate_modules}
# 
# haRow_multiomics <- HeatmapAnnotation(df = multiomics_ann,
#                                     which = "row",
#                                     col = list(Cluster = factor(cluster_cols, levels = cluster_cols, ordered = TRUE),
#                                                Type = assay_cols,
#                                                Set = c("shared" = "black", ligand_cols)))
# 
# hm_multiomics <- Heatmap(multiomics_data_dm,
#                        name = "lfc rr",
#                        column_title = "Multiomics features",
#                        column_title_gp = gpar(fontsize = 12),
#                        col = colorRamp2(c(-1.5, 0, 1.5), c("#2166AC", "white", "#B2182B")),
#                        cluster_rows = FALSE,
#                        cluster_row_slices = FALSE, 
#                        cluster_columns = FALSE,
#                        row_split = multiomics_ann$Cluster,
#                        row_gap = unit(2, "mm"),
#                        row_title = " ",
#                        show_row_names = FALSE,
#                        row_names_gp = gpar(fontsize = 8),
#                        column_names_gp = gpar(fontsize = 10),
#                        left_annotation = haRow_multiomics,
#                        use_raster = FALSE)
# hm_multiomics
# 
# hm_multiomics_mean <- Heatmap(multiomics_data_mean_dm,
#                             name = "lfc ",
#                             column_title = "Multiomics k=18 cluster means",
#                             column_title_gp = gpar(fontsize = 12),
#                             col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
#                             cluster_rows = FALSE,
#                             cluster_row_slices = FALSE,
#                             cluster_columns = FALSE,
#                             row_gap = unit(2, "mm"),
#                             row_title = " ",
#                             show_row_names = TRUE,
#                             row_names_gp = gpar(fontsize = 8),
#                             column_names_gp = gpar(fontsize = 10),
#                             use_raster = FALSE)
# hm_multiomics_mean

#find correlations between the modules and show cut to combine into 14 modules
cor_dm <- cor(t(multiomics_data_mean_dm),  method = "pearson")
hr = hclust(dist(cor_dm))
clusters = dendextend::cutree(hr, k = 14)

cor_hm <- Heatmap(cor_dm,
                  column_title = "Module correlations",
                   name = "pearson\ncorrelation",
                  row_title = " ",
                  row_names_gp = gpar(fontsize = 10),
                  split = clusters,
                  column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_hm

if(create_pdfs) {
  pdf("../multiomics_analysis/plots/MDD_multiomics_module_module_correlations_SF6b.pdf",
      width = 9, height = 8)
  res <- print(cor_hm)
  res <- dev.off()
}

# if(write_csvs) {
#   write.csv(cor_dm, "../tables/MDD_multiomics_module_module_correlations.csv")
# }

```


```{r reduce_to_14_clusters}
#####
#Use the multiomics features data and the cluster assignments to
#create the multiomics_14 dataset

multiomics_14_clusters <-c("1", "2+8", "3", "4", "5+6", "7", "9+18", "10", "11","12","13","14+16", "15","17")
cluster14_cols <- c(structure(RColorBrewer::brewer.pal(12, "Paired")[1:12], names = multiomics_14_clusters[1:12]), c("15" = "#222222", "17" = "#555555"))
                          
multiomics_14_data_ann <- multiomics_data_ann %>%
    mutate(Cluster = as.character(Cluster),
      Cluster = case_when(Cluster %in% c("2","8") ~"2+8",
                          Cluster %in% c("5","6") ~"5+6",
                          Cluster %in% c("9", "18") ~"9+18",
                          Cluster %in% c("14","16") ~"14+16",
                               TRUE ~Cluster),
      Cluster = factor(Cluster, levels = multiomics_14_clusters, ordered = TRUE)) %>%
  as.data.frame()

#create a numeric matrix with feature row names
multiomics_14_dm <- multiomics_14_data_ann %>%
  select(all_of(condition_order)) %>%
  as.matrix() 
rownames(multiomics_14_dm) <- multiomics_14_data_ann$feature_type
multiomics_14_dm <- multiomics_14_dm[!is.nan(multiomics_14_dm[,1]),]

# 
# multiomics_14_features_count <- multiomics_14_data_ann %>%
#   group_by(Cluster) %>%
#   dplyr::count()

#Create a matrix and heatmap of the mean combined18 values
multiomics_14_data_mean <- multiomics_14_data_ann %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

multiomics_14_data_mean_dm <- multiomics_14_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_14_data_mean_dm) <-multiomics_14_data_mean$Cluster

```

```{r 14_module_heatmaps}
#Make heatmap with annotations
# haRow_multiomics_14 <- HeatmapAnnotation(df = multiomics_14_data_ann%>%select(Cluster, Type, Set),
#                                     which = "row",
#                                     col = list(Cluster = factor(cluster14_cols, levels = cluster14_cols, ordered = TRUE),
#                                                Type = assay_cols,
#                                                Set = c("shared" = "white", ligand_cols)))
# 
# hm_multiomics_14 <- Heatmap(multiomics_14_dm,
#                        name = "lfc rr",
#                        column_title = "multiomics_14 features",
#                        column_title_gp = gpar(fontsize = 12),
#                        col = colorRamp2(c(-1.5, 0, 1.5), c("#2166AC", "white", "#B2182B")),
#                        cluster_rows = FALSE,
#                        cluster_row_slices = FALSE, 
#                        cluster_columns = FALSE,
#                        row_split = multiomics_14_data_ann$Cluster,
#                        row_gap = unit(2, "mm"),
#                        row_title = " ",
#                        show_row_names = FALSE,
#                        row_names_gp = gpar(fontsize = 8),
#                        column_names_gp = gpar(fontsize = 10),
#                        left_annotation = haRow_multiomics_14,
#                        use_raster = FALSE)
# hm_multiomics_14

###multiomics_14 heatmap with ligand column annotations, RPPA, cycIF and motif callouts
  callouts <- read_csv(paste0(data_path, "MDD_multiomics_heatmap_callouts.csv")) %>%
    filter(selected == "x")

#Annotate callouts with abbreviations of the type 

df <- multiomics_14_data_ann %>%
  #filter(!Type == "RNAseq") %>%
  mutate(feature = str_remove(feature, "_[[:digit:]]_.*[nuc|cytoplasm]"),
         feature = case_when(Type == "RPPA" ~paste(feature, "(R)"),
                             Type == "cycIF" ~paste(feature, "(C)"),
                             Type == "motifs" ~paste(feature, "(A)"),
                             TRUE ~ feature))

RPPA_motif_rows_right <- which(df$feature_type %in% callouts$feature_type)

haRight = rowAnnotation(foo = anno_mark(at = RPPA_motif_rows_right,
                                   labels = df$feature[RPPA_motif_rows_right],
                                   labels_gp = gpar(fontsize = 5)))

hm_multiomics_14_callouts <- Heatmap(multiomics_14_dm,
                       name = "lfc rr",
                       column_title = "multiomics_14 features",
                       column_title_gp = gpar(fontsize = 6),
                       width = ncol(multiomics_14_dm)*unit(4, "mm"),
                       col = colorRamp2(c(-1.5, 0, 1.5), c("#2166AC", "white", "#B2182B")),
                       cluster_rows = FALSE,
                       cluster_row_slices = FALSE, 
                       cluster_columns = FALSE,
                       row_split = multiomics_14_features_ann$Cluster,
                       row_gap = unit(3, "mm"),
                       #row_title = " ",
                       show_row_names = FALSE,
                       row_names_gp = gpar(fontsize = 12),
                       column_names_gp = gpar(fontsize = 6),
                     # left_annotation = haLeft,
                       right_annotation = haRight,
                       #top_annotation = haCol_multiomics_14,
                       use_raster = FALSE)
hm_multiomics_14_callouts

###
if(create_pdfs){
  pdf("../plots/MDD_integrated_multiomics_14_callouts_heatmap.pdf",height = 10) 
  draw(hm_multiomics_14_callouts)
  res <- dev.off()
} 
# 
# hm_multiomics_14_mean <- Heatmap(multiomics_14_data_mean_dm,
#                               name = "lfc ",
#                               column_title = "multiomics_14 cluster means",
#                               column_title_gp = gpar(fontsize = 12),
#                               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
#                               cluster_rows = FALSE,
#                               cluster_row_slices = FALSE,
#                               cluster_columns = FALSE,
#                               row_gap = unit(2, "mm"),
#                               row_names_side = "left",
#                               row_title = " ",
#                               show_row_names = TRUE,
#                               row_names_gp = gpar(fontsize = 12),
#                               column_names_gp = gpar(fontsize = 10),
#                               use_raster = FALSE)
# draw(hm_multiomics_14_mean, heatmap_legend_side = "left")
# 
# mean_highlight_thresh <- 0.5
# hm_multiomics_14_mean_obj <- Heatmap(multiomics_14_data_mean_dm,
#                             name = "lfc ",
#                             column_title = "multiomics_14 cluster means",
#                             column_title_gp = gpar(fontsize = 12),
#                             col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
#                             cluster_rows = FALSE,
#                             cluster_row_slices = FALSE,
#                             cluster_columns = FALSE,
#                             row_gap = unit(2, "mm"),
#                             show_row_names = TRUE,
#                             row_names_gp = gpar(fontsize = 8),
#                             column_names_gp = gpar(fontsize = 10),
#                             cell_fun = function(j, i, x, y, width, height, fill) {
#         if(abs(multiomics_14_data_mean_dm[i, j]) > mean_highlight_thresh)
#             grid.text(sprintf("%.1f", multiomics_14_data_mean_dm[i, j]), x, y, gp = gpar(fontsize = 12))},
#                             use_raster = FALSE)
#hm_multiomics_14_mean <- draw(hm_multiomics_14_mean_obj)
#multiomics_14_cluster_order <- row_order(hm_multiomics_14_mean)

#Compare the consensus clustering to the previous single clustering
path <- "../multiomics_analysis/data/MDD_combined14_single_tables.xlsx"
multiomics_combined14_single <- path %>% 
  readxl::excel_sheets() %>% 
  set_names() %>% 
  map_dfr(readxl::read_excel, path = path) %>%
  mutate(Cluster = str_remove(Cluster, "module_")) %>%
  arrange(feature_type)

#convert to a clue partition
multiomics_combined14_single_partition<-multiomics_combined14_single %>%
  pull(Cluster) %>%
  set_names(multiomics_combined14_single$feature_type) %>%
  as.cl_partition()

#convert consensus clustering to a clue hard partition
multiomics_14_partition <- multiomics_14_data_ann %>%
   arrange(feature_type) %>%
  pull(Cluster) %>%
  set_names(multiomics_14_data_ann$feature_type) %>%
  as.cl_partition()

single_to_consensus_agreement <- cl_agreement(multiomics_combined14_single_partition, multiomics_14_partition, method = "FM")

#calculate means for single clustering
multiomics_14_single_features_data_mean<- read_csv(paste0(data_path, "MDD_combined14_single_mean.csv"))
multiomics_14_single_features_data_mean_dm <- multiomics_14_single_features_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_14_single_features_data_mean_dm) <- multiomics_14_single_features_data_mean$Cluster
single_to_consensus_cors <- cor(t(multiomics_14_data_mean_dm), t(multiomics_14_single_features_data_mean_dm), method = "pearson")

hm_single_to_consensus_cors <- Heatmap(single_to_consensus_cors, 
                                            column_title = "single to consensus clustering correlations",
                                            name = "pearson\ncorrelations ",
                                            row_names_gp = gpar(fontsize = 5),
                                            column_names_gp = gpar(fontsize = 5),
                                            col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
hm_single_to_consensus_cors

#find set differences between single seed modules 2 and 9 and consensus modules 9 and 10 respectively.
cluster2_single <- multiomics_combined14_single %>%
  filter(Cluster == "2")
cluster4_consensus <- multiomics_14_data_ann %>%
  filter(Cluster == "4")
cluster2_inters <- intersect(cluster2_single$feature_type, cluster4_consensus$feature_type)
cluster2_dif <- setdiff(cluster2_single$feature_type, cluster4_consensus$feature_type)
cluster4_dif <- setdiff(cluster4_consensus$feature_type,cluster2_single$feature_type)

multiomics_combined14_single_2_9 <- multiomics_combined14_single %>%
  filter(Cluster %in% c("2","9")) %>%
  rename(Single_Clustering = Cluster) %>%
  select(feature_type, Single_Clustering)
multiomics_14_data_ann_4_10 <- multiomics_14_data_ann %>%
  filter(Cluster %in% c("4", "10")) %>%
  rename(Consensus_Clustering =Cluster) %>%
  mutate(Consensus_Clustering = as.character(Consensus_Clustering)) %>%
select(Consensus_Clustering, feature_type)
clustering_module_comparisons <- multiomics_combined14_single_2_9 %>%
 full_join(multiomics_14_data_ann_4_10, by = "feature_type") %>%
  mutate(Both = case_when(!is.na(Single_Clustering)&!is.na(Consensus_Clustering) ~ "x",
                          TRUE ~ ""))
 write_csv(clustering_module_comparisons, "../multiomics_analysis/tables/MDD_multiomics_cl1_cl9_comparisons.csv")

 clustering_module_comparisons_counts <- clustering_module_comparisons %>%
   group_by(Single_Clustering, Consensus_Clustering, Both) %>%
   count()
 write_csv(clustering_module_comparisons_counts, "../multiomics_analysis/tables/MDD_multiomics_cl1_cl9_comparisons_counts.csv")

 #Remove the cluster assignments to keep the code similar to the 
#other dataset processing
#Remove RNAseq and label rows

#####
if(create_pdfs){
  pdf("..//multiomics_analysis/plots/MDD_multiomics_heatmap_F5a.pdf",height = 10) 
  draw(hm_multiomics_14_callouts)
  # draw(hm_multiomics_14_mean)
  # draw(hm_single_to_consensus_cors)
  res <- dev.off()
} 

```

```{r writeTables}
selectFeature <- function(df){
  #browser()
  res <- df %>%
    select(feature, Type) %>%
    mutate(analyte = str_remove(feature, "_.*")) %>%
    rename(assay = Type) %>%
    select(feature, analyte, assay, time, symbol)
return(res)
}

selectDerivedFeature <- function(df){
  #browser()
  res <- df %>%
    select(feature, FDR, Type)
return(res)
}

multiomics_14 <- multiomics_14_data_ann %>%
  arrange(Cluster, Type) %>%
  mutate(Cluster = paste0("module_", Cluster),
         Cluster = factor(Cluster, levels = unique(Cluster))) %>%
  split(.$Cluster) 

res <- write_xlsx(multiomics_14,
                  path = "../multiomics_analysis/tables/MDD_multiomics_14.xlsx")

res <- multiomics_14_data_ann %>%
  arrange(Cluster, Type) %>%
  mutate(Cluster = paste0("module_", Cluster),
         Cluster = factor(Cluster, levels = unique(Cluster))) %>%
  write_csv(file = "../multiomics_analysis/tables/MDD_multiomics_14.csv")

#res <- write_csv(multiomics_14_data_mean, "../multiomics_analysis/tables/MDD_multiomics_14_mean.csv")

```

### Ligand heatmaps

```{r ligand_heatmaps}

#make a heatmap of the unique features
#Create data and metadata matrices and annotations

unique_features_data_ann <- unique_features_data %>%
  mutate(Ligand = str_remove(Cluster, "_.*"),
         Ligand = factor(Ligand,  ordered = TRUE),
         Direction = factor(Direction, ordered = TRUE),
         Cluster = factor(Cluster, ordered = TRUE)) 

unique_features_ann <- unique_features_data_ann %>%
  select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_unique <- HeatmapAnnotation(df = unique_features_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

unique_features_data_dm<- unique_features_data_ann %>%
  select(all_of(condition_order)) %>%
  as.matrix()
rownames(unique_features_data_dm) <- unique_features_data_ann$feature

unique_features_data_mean <- unique_features_data_ann %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

unique_features_count <- unique_features_data_ann %>%
  group_by(Cluster) %>%
  count()

unique_features_data_mean_dm <- unique_features_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(unique_features_data_mean_dm) <- paste0(unique_features_data_mean$Cluster, "_",unique_features_count$n)

hm_unique <- Heatmap(unique_features_data_dm,
                     name = "lfc rr",
                     column_title = "Ligand-specific features",
                     column_title_gp = gpar(fontsize = 12),
                     col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
                     cluster_rows = FALSE,
                     cluster_row_slices = TRUE, 
                     cluster_columns = FALSE,
                     row_split = unique_features_data_ann$Cluster,
                     row_gap = unit(2, "mm"),
                     row_title = " ",
                     show_row_names = FALSE,
                     row_names_gp = gpar(fontsize = 2),
                     column_names_gp = gpar(fontsize = 10),
                     left_annotation = haRow_unique,
                     use_raster = FALSE)
hm_unique

hm_unique_mean <- Heatmap(unique_features_data_mean_dm,
                          name = "lfc ",
                          column_title = "Ligand-specific means",
                          column_title_gp = gpar(fontsize = 12),
                          col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
                          cluster_rows = TRUE,
                          cluster_row_slices = FALSE,
                          cluster_columns = FALSE,
                          row_gap = unit(2, "mm"),
                          row_title = " ",
                          show_row_names = TRUE,
                          row_names_gp = gpar(fontsize = 8),
                          column_names_gp = gpar(fontsize = 10),
                          use_raster = FALSE)
hm_unique_mean

###Heatmap of ligand-specific RPPA48 hour features with rownames
unique_features_RPPA_48 <- unique_features_data_ann %>%
  filter(Type == "RPPA") %>%
  select(-contains("_24"))

unique_features_RPPA_48_dm <- unique_features_RPPA_48 %>%
  select(where(is.numeric)) %>%
  as.matrix()
rownames(unique_features_RPPA_48_dm) <- unique_features_RPPA_48$feature

unique_features_RPPA_48_ann <- unique_features_RPPA_48 %>%
  select(Ligand, Type, Direction) %>%
  data.frame()

haRow_unique_RPPA_48 <- HeatmapAnnotation(df = unique_features_RPPA_48_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_unique_RPPA_48 <- Heatmap(unique_features_RPPA_48_dm,
                     name = "lfc rr",
                     column_title = "Ligand-specific RPPA 48H, ordered",
                     column_title_gp = gpar(fontsize = 12),
                     col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
                     cluster_rows = FALSE,
                     cluster_row_slices = FALSE, 
                     cluster_columns = FALSE,
                     row_split = unique_features_RPPA_48$Cluster,
                     row_gap = unit(2, "mm"),
                     row_title = " ",
                     show_row_names = TRUE,
                     row_names_gp = gpar(fontsize = 7),
                     column_names_gp = gpar(fontsize = 10),
                     left_annotation = haRow_unique_RPPA_48,
                     use_raster = FALSE)
hm_unique_RPPA_48

###48 hour motifs

###Heatmap of ligand-specific motifs48 hour features with rownames
unique_features_motifs_48 <- unique_features_data_ann %>%
  filter(Type == "motifs") %>%
  select(-contains("_24"))

unique_features_motifs_48_dm <- unique_features_motifs_48 %>%
  select(where(is.numeric)) %>%
  as.matrix()
rownames(unique_features_motifs_48_dm) <- unique_features_motifs_48$feature

unique_features_motifs_48_ann <- unique_features_motifs_48 %>%
  select(Ligand, Type, Direction) %>%
  data.frame()

haRow_unique_motifs_48 <- HeatmapAnnotation(df = unique_features_motifs_48_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_unique_motifs_48 <- Heatmap(unique_features_motifs_48_dm,
                     name = "lfc rr",
                     column_title = "Ligand-specific motifs 48H, ordered",
                     column_title_gp = gpar(fontsize = 12),
                     col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
                     cluster_rows = FALSE,
                     cluster_row_slices = FALSE, 
                     cluster_columns = FALSE,
                     row_split = unique_features_motifs_48$Cluster,
                     row_gap = unit(2, "mm"),
                     row_title = " ",
                     show_row_names = TRUE,
                     row_names_gp = gpar(fontsize = 7),
                     column_names_gp = gpar(fontsize = 10),
                     left_annotation = haRow_unique_motifs_48,
                     use_raster = FALSE)
hm_unique_motifs_48

###48 hour cycIF
###Heatmap of ligand-specific cycIF48 hour features with rownames
unique_features_cycIF_48 <- unique_features_data_ann %>%
  filter(Type == "cycIF") %>%
  select(-contains("_24"))

unique_features_cycIF_48_dm <- unique_features_cycIF_48 %>%
  select(where(is.numeric)) %>%
  as.matrix()
rownames(unique_features_cycIF_48_dm) <- unique_features_cycIF_48$feature

unique_features_cycIF_48_ann <- unique_features_cycIF_48 %>%
  select(Ligand, Type, Direction) %>%
  data.frame()

haRow_unique_cycIF_48 <- HeatmapAnnotation(df = unique_features_cycIF_48_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_unique_cycIF_48 <- Heatmap(unique_features_cycIF_48_dm,
                     name = "lfc rr",
                     column_title = "Ligand-specific cycIF 48H, ordered",
                     column_title_gp = gpar(fontsize = 12),
                     col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
                     cluster_rows = FALSE,
                     cluster_row_slices = FALSE, 
                     cluster_columns = FALSE,
                     row_split = unique_features_cycIF_48$Cluster,
                     row_gap = unit(2, "mm"),
                     row_title = " ",
                     show_row_names = TRUE,
                     row_names_gp = gpar(fontsize = 7),
                     column_names_gp = gpar(fontsize = 10),
                     left_annotation = haRow_unique_cycIF_48,
                     use_raster = FALSE)
hm_unique_cycIF_48

#No RNAseq and motifs
df <- unique_features_data_ann %>%
  filter(!Type %in% c("RNAseq", "motifs"))

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_noRNAseq <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, no RNAseq, no motifs, row-clustered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = TRUE,
               cluster_row_slices = TRUE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 5),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_noRNAseq

# #GCP only version
df <- unique_features_data_ann %>%
  filter(Type == "GCP")

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature


df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_GCP <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, GCP, row-clustered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = TRUE,
               cluster_row_slices = TRUE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_GCP

#cycIF only version
df <- unique_features_data_ann %>%
  filter(Type == "cycIF")

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))


hm_cycIF <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, cycIF, row-clustered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = TRUE,
               cluster_row_slices = TRUE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_cycIF

#RPPA only version
df <- unique_features_data_ann %>%
  filter(Type == "RPPA")

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_RPPA <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, RPPA, row-clustered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = TRUE,
               cluster_row_slices = TRUE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_RPPA

if(create_pdfs){
  pdf("../plots/MDD_integrated_rr_ligand_heatmap.pdf")
  draw(hm_unique)
  draw(hm_unique_RPPA_48)
   draw(hm_unique_motifs_48)
     draw(hm_unique_cycIF_48)
     draw(hm_noRNAseq)
  draw(hm_GCP)
  draw(hm_cycIF)
  draw(hm_RPPA)
  res <- dev.off()
}


# #heatmaps sorted by ligand and direction
#no RNAseq and motifs
df <- unique_features_data_ann %>%
  filter(!Type %in% c("RNAseq", "motifs")) %>%
  arrange(Set, Direction)

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_noRNAseq <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, no RNAseq, motifs, ordered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = FALSE,
               cluster_row_slices = FALSE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_noRNAseq

# #GCP version
df <- unique_features_data_ann %>%
  filter(Type == "GCP") %>%
  arrange(Set, Direction)

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_GCP <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, GCP, ordered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = FALSE,
               cluster_row_slices = FALSE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_GCP

# #cycIF version
df <- unique_features_data_ann %>%
  filter(Type == "cycIF") %>%
  arrange(Set, Direction)

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_cycIF <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, cycIF, ordered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = FALSE,
               cluster_row_slices = FALSE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_cycIF

#RPPA version
df <- unique_features_data_ann %>%
  filter(Type == "RPPA") %>%
  arrange(Set, Direction)

df_dm<- df %>%
    select(all_of(condition_order)) %>%
  as.matrix()
rownames(df_dm) <- df$feature

df_ann <- df %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))

hm_RPPA <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, RPPA, ordered",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = FALSE,
               cluster_row_slices = FALSE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 8),
               column_names_gp = gpar(fontsize = 10),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_RPPA

if(create_pdfs){
  pdf("../plots/../plots/MDD_integrated_rr_ligand_heatmap_ordered.pdf")
  draw(hm_noRNAseq)
  draw(hm_GCP)
  draw(hm_cycIF)
  draw(hm_RPPA)
  res <- dev.off()
}

####
#Create a matrix and heatmap of the mean Ligand-specific values

if(create_pdfs){
  pdf("../plots/../plots/MDD_ligand_heatmaps.pdf") 
  draw(hm_unique)
  draw(hm_unique_mean)
  res <- dev.off()
} 

```


### UMAPs

```{r UMAPAnalysis, fig.width=5.75, fig.height=4.75}

#Generate a 2d scatterplot of the UMAP dimensions and color by clusters, assay type

#Create annotation values
dm <- multiomics_14_data_ann %>%
  dplyr::select(-feature, -Cluster, -Type, -feature_type, -Set) %>%
  as.matrix()
rownames(dm) <- multiomics_14_data_ann$feature_type

custom_settings = umap.defaults
custom_settings$random_state <- 42
custom_settings$n_epochs = 500
custom_settings$n_neighbors = 25

if(file.exists(paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP.rda"))){
  load(paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP.rda"))
} else {
df_UMAP <- umap(dm, config = custom_settings, na.rm = TRUE)$layout %>%
  data.frame(feature_type = rownames(dm)) %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
  left_join(multiomics_14_data_ann, by = "feature_type") 
save(df_UMAP, file = paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP.rda"))
}

  p <- ggplot(df_UMAP, aes(x = UMAP_1,
                           y = UMAP_2,
                           colour = factor(Cluster))) +
    geom_point(size = 1, alpha = .6, shape = 16) +
    scale_color_manual(values = cluster14_cols) +
    labs(title = paste("UMAP embedding colored by cluster, multiomics_14"),
         colour = "Cluster") +
    guides(colour = guide_legend(override.aes = list(alpha = 1, size = 4))) +
    theme_bw()+
    theme(axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line =element_blank())
  p
  
  if(create_pdfs) {
    pdf("../plots/MDD_multiomics_14_umap_main.pdf")    
    print(p)
    res <- dev.off()
  } 
```

There are `r dim(dm)[1]` dots in the 2d UMAP that represent the significant differentially expressed features relative to T0 in the multiomics_14-features integrated matrix.  


```{r UMAPAnalysis2, fig.width=7.5, fig.height=4}
p <- ggplot(data = df_UMAP[df_UMAP$Type=="RNA",], aes(x = UMAP_1,
                                                      y = UMAP_2,
                                                      colour = factor(Type))) +
  geom_point(size = .8, alpha = .6) +
  scale_colour_manual(values = assay_cols)+
  labs(title = paste("UMAP embedding colored by assay, multiomics_14"),
       colour = "Assay") +
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 4))) +
  theme_bw()+
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line =element_blank())
p <- p + geom_point(data = df_UMAP[!df_UMAP$Type=="RNA",],size = .8, alpha = .8)
p

if(create_pdfs) {
  pdf("../plots/MDD_ligands_multiomics_14_umap_type.pdf") 
      print(p)
  res <- dev.off()
}

```

```{r highlightAssay, fig.width=2.5, fig.height=2.5}

p_list <- lapply(unique(df_UMAP$Type), function(assay){
  
  df <- df_UMAP
  
  p <- ggplot(data = df[!df$Type==assay,], 
              aes(x = UMAP_1,
                  y = UMAP_2,
                  color = Type)) +
    geom_point(color = "gray",size = 1, alpha = .4, shape = 16) +
    scale_colour_manual(values = assay_cols) +
    annotate("text",x = -2, y = -6, label = assay) +
    scale_size_area(limits = c(0, 30)) +
    labs()+
    guides(colour = FALSE,
           size = FALSE)+
    theme_bw()+
    theme(axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line =element_blank())
  p <- p + geom_point(data = df[df$Type==assay,], 
                      size = 1,
                      alpha = .9,
                      shape = 16) 
  return(p)
  
})

res <- map(p_list, print)

if(create_pdfs) {
  pdf("../plots/MDD_ligands_multiomics_14_umap_type_highlights.pdf") 
  res <- map(p_list, print)
  res <- dev.off()
}
```


```{r UMAPbyCondition, fig.width=2.5, fig.height=2.5, eval = TRUE}


df <- df_UMAP %>%
  rename("BMP2_EGF_24" = "BMP2+EGF_24",
         "IFNG_EGF_24" =  "IFNG+EGF_24",
         "TGFB_EGF_24" = "TGFB+EGF_24",
         "BMP2_EGF_48" = "BMP2+EGF_48",
         "IFNG_EGF_48" =  "IFNG+EGF_48",
         "TGFB_EGF_48" = "TGFB+EGF_48")
df_UMAP_conditions <- colnames(df)[str_detect(colnames(df_UMAP),"PBS|HGF|OSM|EGF|BMP2|IFNG|TGFB")] %>%
  str_replace("[+]","_")

p_list <- lapply(df_UMAP_conditions, function(condition){
  ligand <- str_remove(condition, "_[[:digit:]]*$") %>%
    str_replace("_", "+")

  highlights <- df %>%
    filter(Set == ligand,
           feature_type %in% c()) 
  
  p_base <- ggplot(df, aes_string(x = "UMAP_1",
                             y = "UMAP_2",
                             colour = condition)) +
    geom_point(size = 2, alpha = .8, shape = 16) +
    annotate("text",x = -2, y = -6, label = condition)+
    scale_color_gradientn(colours = colorRampPalette(rev(brewer.pal(n = 7, name =
                                                                      "RdBu")))(100)) +
    labs()+
    guides(colour = FALSE)+
    theme_bw()+
    theme(text = element_text(size = 1),
          axis.title = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black"))

  p <- p_base + geom_text(data = highlights, aes(label = feature_type), color = "black") +
    geom_point(data = highlights, color = "black")
  return(p)
})
res <- map(p_list, print)
if(create_pdfs) {
  pdf("../plots/MDD_integrated_multiomics_14_umap_by_condition.pdf")
  res <- map(p_list, print)
  res <- dev.off()
}

#Now create a version of RNAseq features only
dm <- multiomics_14_data_ann %>%
  filter(Type == "RNAseq") %>%
  dplyr::select(-feature, -Cluster, -Type, -feature_type, -Set) %>%
  as.matrix()
rownames(dm) <- multiomics_14_data_ann %>%
    filter(Type == "RNAseq") %>%
  pull(feature_type)

custom_settings = umap.defaults
custom_settings$random_state <- 42
custom_settings$n_epochs = 500
custom_settings$n_neighbors = 25

if(file.exists(paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP_RNAseq.rda"))){
  load(paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP_RNAseq.rda"))
} else {
df_UMAP <- umap(dm, config = custom_settings, na.rm = TRUE)$layout %>%
  data.frame(feature_type = rownames(dm)) %>%
  rename(UMAP_1 = X1,
         UMAP_2 = X2) %>%
  left_join(multiomics_14_data_ann, by = "feature_type") 
save(df_UMAP, file = paste0("../integrated_analysis/MDD_",cluster_method,"_multiomics_14_UMAP_RNAseq.rda"))
}

df <- df_UMAP %>%
  rename("BMP2_EGF_24" = "BMP2+EGF_24",
         "IFNG_EGF_24" =  "IFNG+EGF_24",
         "TGFB_EGF_24" = "TGFB+EGF_24",
         "BMP2_EGF_48" = "BMP2+EGF_48",
         "IFNG_EGF_48" =  "IFNG+EGF_48",
         "TGFB_EGF_48" = "TGFB+EGF_48")
df_UMAP_conditions <- colnames(df)[str_detect(colnames(df_UMAP),"PBS|HGF|OSM|EGF|BMP2|IFNG|TGFB")] %>%
  str_replace("[+]","_")

p_list <- lapply(df_UMAP_conditions, function(condition){
  ligand <- str_remove(condition, "_[[:digit:]]*$") %>%
    str_replace("_", "+")

  highlights <- df %>%
    filter(Set == ligand,
           feature_type %in% c()) 
  
  p_base <- ggplot(df, aes_string(x = "UMAP_1",
                             y = "UMAP_2",
                             colour = condition)) +
    geom_point(size = 2, alpha = .8, shape = 16) +
    annotate("text",x = -2, y = -6, label = condition)+
    scale_color_gradientn(colours = colorRampPalette(rev(brewer.pal(n = 7, name =
                                                                      "RdBu")))(100)) +
    labs()+
    guides(colour = FALSE)+
    theme_bw()+
    theme(text = element_text(size = 1),
          axis.title = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(colour = "black"))

  p <- p_base + geom_text(data = highlights, aes(label = feature_type), color = "black") +
    geom_point(data = highlights, color = "black")
  return(p)
})
res <- map(p_list, print)
if(create_pdfs) {
  pdf("../plots/MDD_integrated_multiomics_14_umap_RNAseq_by_condition.pdf")
  res <- map(p_list, print)
  res <- dev.off()
}

```

### phenotype analysis

```{r correlate_to_phenotypes}

#start with level 3 IF data from synapse
if_df <- read_csv("../Immunofluorescence/Data/MDD_IF_Level3.csv",skip = 1) %>%
  rename("feature" = "specimenName") %>%
  filter(str_detect(feature, "_")) %>%
  select(matches("feature|24_C2|48_C2")) %>%
  mutate(across(-feature, as.numeric)) %>%
  pivot_longer(cols = matches("_C2_")) %>%
  mutate(condition = str_remove(name, "_C2_.$")) %>%
  mutate(condition = str_replace(condition, "BMP2","BMP2+EGF"),
         condition = str_replace(condition, "IFNG","IFNG+EGF"),
         condition = str_replace(condition, "TGFB","TGFB+EGF")) %>%
  group_by(feature, condition) %>%
  summarise(value = mean(value), .groups = "drop") %>%
  ungroup() %>%
  pivot_wider(names_from = condition, values_from = value) %>%
  select(feature, all_of(condition_order))

if_df_migration <- read_csv("../Immunofluorescence/Data//MDD_migration_distance_lineage_T0.csv") %>%
  filter(thour %in%c(24, 48),
         !treatment == "PBS") %>%
  mutate(condition = paste0(treatment,"_",thour),
         feature = "migration_distance")  %>%
  select(-treatment, -thour) %>%
  pivot_wider(names_from = condition, values_from = lineage_length_mean) %>%
  bind_rows(if_df) 

phenotype_data_ann <- if_df_migration %>%
  filter(str_detect(feature, "Well_Cell_Count|Mean_Cells_per_Cluster|Normalized_Second_Neighbor_Dist|DNA2n_Proportion|Edu_Positive_Proportion|MeanIntensity_KRT5|migration_distance")) %>%
  pivot_longer(cols = -feature, names_to = "condition", values_to = "phenotype_value") %>%
  group_by(feature) %>%
  mutate(phenotype_value = scale(phenotype_value)) %>%
  ungroup()

module_phenotype_data_ann <- multiomics_14_data_mean %>%
  pivot_longer(cols = -Cluster, names_to = "condition", values_to = "module_value") %>%
  mutate(ligand = str_remove(condition, "_.*"),
         time = str_remove(condition, ".*_") %>% as.integer) %>%
  rename(module = Cluster) %>%
  inner_join(phenotype_data_ann, by = "condition")

df <- module_phenotype_data_ann %>%
  filter(module == "10",
         feature == "Edu_Positive_Proportion")

p_mod10_edu <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm") +
  geom_point(aes(color = ligand, size = time), alpha = .9) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 10 EdU positive proportion vs module",
       x = "Module expression",
       y = "EdU positive proportion") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod10_edu

df <- module_phenotype_data_ann %>%
  filter(module == "9+18",
         feature == "Mean_Cells_per_Cluster")

p_mod9_18_mcc <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm") +
  geom_point(aes(color = ligand, size = time), alpha = .9) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 9+18 Mean_Cells_per_Cluster proportion vs module",
       x = "Module expression",
       y = "Mean cells per cluster") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod9_18_mcc

df <- module_phenotype_data_ann %>%
  filter(module == "7",
         feature == "Mean_Cells_per_Cluster")

p_mod7_mcc <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm") +
  geom_point(aes(color = ligand, size = time), alpha = .9) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 7 Mean_Cells_per_Cluster proportion vs module",
       x = "Module expression",
       y = "Mean cells per cluster") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod7_mcc

df <- module_phenotype_data_ann %>%
  filter(module == "2+8",
         feature == "Normalized_Second_Neighbor_Dist")

p_mod2_8_nsnd <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm") +
  geom_point(aes(color = ligand, size = time), alpha = .9) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 2+8 Normalized_Second_Neighbor_Dist proportion vs module",
       x = "Module expression",
       y = "Normalized second neighbor dist") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod2_8_nsnd

if(create_pdfs) {
  pdf("../multiomics_analysis/plots/MDD_multiomics_F5d.pdf")
  res <- print(p_mod10_edu)
  res <- dev.off()
  pdf("../multiomics_analysis/plots/MDD_multiomics_F5e.pdf")
  res <- print(p_mod9_18_mcc)
  res <- dev.off() 
  pdf("../multiomics_analysis/plots/MDD_multiomics_F5f.pdf")
  res <- print(p_mod7_mcc)
  res <- dev.off() 
  pdf("../multiomics_analysis/plots/MDD_multiomics_F5g.pdf")
  res <- print(p_mod2_8_nsnd)
  res <- dev.off()
}

```


### cluster analysis


```{r featureTypes, fig.width=8, fig.height=4}
# #create a bar chart of the shared feature types
# df <- shared_features_data %>%
#   #drop_na() %>%
#   select(feature, Type) %>%
#   distinct()
# 
# p <- ggplot(df, aes(x = Type, fill = Type)) +
#   geom_bar() +
#   coord_cartesian(ylim = c(0, 10000)) +
#   scale_fill_manual(values = assay_cols)+
#   labs(title = paste("Type distribution of the",length(unique(paste0(df$feature))), "features at unique timepoints in the shared-features MDD integrated dataset"),
#        x ="Feature type") +
#   theme_bw()+
#   theme(axis.text.x =  element_text(angle = 90),
#     axis.ticks = element_blank(),
#     panel.border = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     axis.line =element_blank())
# 
# p
# 
# p_log <- ggplot(df, aes(x = Type, fill = Type)) +
#   geom_bar() +
#   scale_y_log10() +
#   scale_fill_manual(values = assay_cols)+
#   labs(title = paste("Type distribution of the",length(unique(paste0(df$feature))), "shared features at unique timepoints in the MDD integrated dataset"),
#        x ="Feature type") +
#   theme_bw()+
#   theme(axis.text.x =  element_text(angle = 90),
#     axis.ticks = element_blank(),
#     panel.border = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     axis.line =element_blank())
# 
# p_log
# if(create_pdfs){
#   pdf("../plots/MDD_shared_feature_type_distribution.pdf")
#   print(p)  
#   res <- dev.off()
#   pdf("../plots/MDD_shared_feature_type_distribution_log_scale.pdf")
#   print(p_log)
#   res <- dev.off()
# } 

#create a bar chart of the multiomics_14 feature types
df <- multiomics_14_data_ann %>%
  #drop_na() %>%
  select(feature, Type) %>%
  distinct()


p_log <- ggplot(df, aes(x = Type, fill = Type)) +
  geom_bar() +
  scale_y_log10() +
  scale_fill_manual(values = assay_cols)+
  labs(title = "Multiomics assay features",
       x ="",
       y = "Feature Count (log scaled)") +
  theme_bw()+
  theme(axis.text.x =  element_text(angle = 90),
    axis.ticks = element_blank(),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank())

p_log
if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_multiomics_SF6a.pdf")
  print(p_log)
  res <- dev.off()
} 

#bar chart of the multiomics_14 shared andligands
df <- multiomics_14_data_ann %>%
  count(Set, Cluster) %>%
  mutate(Cluster = factor(Cluster, levels = names(cluster14_cols), ordered = TRUE),
         Set = factor(Set, levels = c(ligand_order, "shared"), ordered = TRUE))

feature_dist <- ggplot(df, aes(x = Set, y = n)) +
  geom_col() +
  scale_fill_manual(values = assay_cols)+
  labs(title = paste("Distribution of the",length(unique(paste0(multiomics_14_data_ann$feature))), "multiomics_14 features"),
       x ="Type") +
  theme_bw()+
  theme(axis.text.x =  element_text(angle = 90),
    axis.ticks = element_blank(),
    #panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank()) +
  facet_wrap(~Cluster, ncol = 1, scales = "free_y")
feature_dist

feature_dist_fill <- ggplot(df, aes(x = Set, y = n, fill = Set)) +
  geom_col() +
  scale_fill_manual(values = c(ligand_cols, "shared" = "black")) +
  labs(title = paste("Distribution of the",length(unique(paste0(multiomics_14_data_ann$feature))), "multiomics_14 features"),
       x ="Type") +
  theme_bw()+
  theme(axis.text.x =  element_text(angle = 90),
    axis.ticks = element_blank(),
    #panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank()) +
  facet_wrap(~Cluster, ncol = 1, scales = "free_y")
feature_dist_fill

if(create_pdfs){
  pdf("../plots/MDD_integrated_multiomics_14_cluster_feature_dist.pdf", width = 3, height = 14)
  print(feature_dist)
  print(feature_dist_fill)
  res <- dev.off()
}

if(write_csvs){
  df_wide <- df %>%
    pivot_wider(names_from = Set, values_from = n)
  res <- write_csv(df_wide, path = "../tables/MDD_integrated_feature_distribution.csv")
}

# df <- multiomics_14_data_ann %>%
#   pivot_longer(cols = matches("_[[:digit:]]"), names_to = "condition") %>%
#   mutate(ligand = str_remove(condition, "_.*"))
# p <- ggplot(df, aes(x = condition, fill = ligand)) +
#   geom_bar() +
#   facet_wrap(~Type, scale = "free_y")
# p

```


```{r gapStatistics, fig.width=4, fig.height=4, eval = TRUE}

plot_gap_custom <- function(x, dataset_name = "") {
  gstab = data.frame(x$Tab, k = seq_len(nrow(x$Tab)))
  p <- ggplot(gstab, aes(k, gap)) + 
    geom_line() +
    geom_errorbar(aes(ymax = gap + SE.sim,
                      ymin = gap - SE.sim), width=0.1) +
    geom_point(size = .5, col=  "red") +
    labs(title = paste("Gap analysis,",cluster_method, dataset_name)) +
     theme_bw()+
  theme(axis.ticks = element_blank(),
        panel.border = element_blank(),
        axis.line =element_blank(),
        title = element_text(size = 8))
  return(p)
}

dm <- multiomics_data %>%
    select(all_of(condition_order)) %>%
  as.matrix()

if(file.exists(paste0("../integrated_analysis/gss_",cluster_method,"_combined18.rda"))){
  load(paste0("../integrated_analysis/gss_",cluster_method,"_combined18.rda"))
} else {
  gss_combined18 <- clusGap(dm, FUN = pam, pamonce = 5, K.max = 25, B = 100, verbose = FALSE)
  save(gss_combined18, file = paste0("../integrated_analysis/gss_",cluster_method,"_combined18.rda"))
}

  plot_gap_custom(gss_combined18, "combined18")
if(create_pdfs){
  pdf("../plots/MDD_combined18_gap_cluster.pdf")
  p <- plot_gap_custom(gss_combined18)
  print(p)
  res <- dev.off()
}

#Run gap analysis on RNAseq only  
  dm <- multiomics_data %>%
    filter(Type == "RNAseq") %>%
    select(all_of(condition_order)) %>%
    as.matrix()
  
  if(file.exists(paste0("../integrated_analysis/gss_RNAseq_",cluster_method,"_combined18.rda"))){
    load(paste0("../integrated_analysis/gss_RNAseq_",cluster_method,"_combined18.rda"))
  } else {
    gss_RNAseq_combined18 <- clusGap(dm, FUN = pam, pamonce = 5, K.max = 25, B = 100, verbose = FALSE)
    save(gss_RNAseq_combined18, file = paste0("../integrated_analysis/gss_RNAseq_",cluster_method,"_combined18.rda"))
  }
  
  plot_gap_custom(gss_RNAseq_combined18, "ombined18_RNAseq")
  if(create_pdfs){
    pdf("../plots/MDD_combined18_RNAseq_gap_cluster.pdf")
    p <- plot_gap_custom(gss_RNAseq_combined18)
    print(p)
    res <- dev.off()
  }
  
# #Shared gap analysis
# dm <- shared_features_data %>%
#     select(all_of(condition_order)) %>%
#   as.matrix()
# 
# if(file.exists(paste0("../integrated_analysis/gss_",cluster_method,"_shared.rda"))){
#   load(paste0("../integrated_analysis/gss_",cluster_method,"_shared.rda"))
# } else {
#   gss_shared <- clusGap(dm, FUN = pam, pamonce = 5, K.max = 25, B = 100, verbose = FALSE)
#   save(gss_shared, file = paste0("../integrated_analysis/gss_",cluster_method,"_shared.rda"))
# }
# 
#   plot_gap_custom(gss_shared, "shared")
# if(create_pdfs){
#   pdf("../plots/MDD_shared_gap_cluster.pdf")
#   p <- plot_gap_custom(gss_shared)
#   print(p)
#   res <- dev.off()
# }
```

```{r assayClusterdistribution, fig.width=8, fig.height=4, eval = TRUE}
# 
# #Shared cluster analysis
# p_hist <- ggplot(shared_features_data_ann, aes(x = as.factor(Cluster), fill = Type)) +
#   geom_bar(position = "dodge") +  
#   scale_fill_manual(values = assay_cols) +
#   labs(title = "Shared cluster analysis",
#        x = "Cluster",
#        fill = "Assay") +
#   facet_wrap(~Type, scales = "free_y") +
#   theme_bw()+
#   theme(#axis.title = element_blank(),
#     #axis.text = element_blank(),
#     axis.ticks = element_blank(),
#     panel.border = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     axis.line =element_blank())
# p_hist
# 
# df <- shared_features_data_ann %>%
#   group_by(Type, Cluster) %>%
#   count() %>%
#   ungroup()
# 
# p <- ggplot(df, aes(x = n, fill = Type)) +
#   geom_density() +  
#   scale_fill_manual(values = assay_cols) +
#   labs(title = "Shared cluster analysis",
#        x = "feature count",
#        fill = "Assay") +
#   facet_wrap(~Type, scales = "free") +
#   theme_bw()+
#   theme(#axis.title = element_blank(),
#     #axis.text = element_blank(),
#     axis.ticks = element_blank(),
#     panel.border = element_blank(),
#     panel.grid.major = element_blank(),
#     panel.grid.minor = element_blank(),
#     axis.line =element_blank())
# p
# 
# if(create_pdfs){
#   pdf("../plots/MDD_shared_cluster_feature_counts.pdf")
#    print(p_hist)
#    print(p)
#   res <- dev.off()
# } 

#Ligand-specific analysis
p_hist <- ggplot(unique_features_data_ann, aes(x = as.factor(Cluster), fill = Type)) +
  geom_bar(position = "dodge") +  
  scale_fill_manual(values = assay_cols) +
  labs(title = "Ligand-specific cluster analysis",
       x = "Cluster",
       fill = "Assay") +
  facet_wrap(~Type, scales = "free_y") +
  theme_bw()+
  theme(#axis.title = element_blank(),
    axis.text.x = element_text(angle = 90),
    axis.ticks = element_blank(),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank())
p_hist

#color bars by ligand 
df <- unique_features_data_ann %>%
  mutate(ligand = str_remove(Cluster, "_.*"))
p_ligand <- ggplot(df, aes(x = as.factor(Cluster), fill = ligand)) +
  geom_bar(position = "dodge") +  
  scale_fill_manual(values = ligand_cols) +
  labs(title = "Ligand-specific cluster analysis",
       x = "Cluster",
       fill = "Ligand") +
  theme_bw()+
    theme(axis.ticks = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines")) +
    facet_wrap(~Type, scales = "free_y")
p_ligand

if(create_pdfs){
  pdf("../plots/MDD_ligand_cluster_feature_counts.pdf", width = 5, height = 3)
   print(p_ligand)
  res <- dev.off()
} 

####
df <- unique_features_data_ann %>%
  group_by(Type, Cluster) %>%
  count() %>%
  ungroup()

p <- ggplot(df, aes(x = n, fill = Type)) +
  geom_density() +  
  scale_fill_manual(values = assay_cols) +
  labs(title = "Ligand-specific analysis",
       x = "feature count",
       fill = "Assay") +
  facet_wrap(~Type, scales = "free") +
  theme_bw()+
  theme(#axis.title = element_blank(),
    axis.text.x = element_text(angle = 0),
    axis.ticks = element_blank(),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank())
p

if(create_pdfs){
  pdf("../plots/MDD_unique_cluster_feature_counts.pdf")
   print(p_hist)
   print(p_ligand)
   print(p)
  res <- dev.off()
} 

#multiomics_14 cluster analysis
p_hist <- ggplot(multiomics_14_data_ann, aes(x = as.factor(Cluster), fill = Type)) +
  geom_bar(position = "dodge") +  
  scale_fill_manual(values = assay_cols) +
  labs(title = "multiomics_14 cluster analysis",
       x = "Cluster",
       fill = "Assay") +
  facet_wrap(~Type, scales = "free_y") +
  theme_bw()+
  theme(#axis.title = element_blank(),
    #axis.text = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines"))
p_hist

if(create_pdfs){
  pdf("../plots/MDD_multiomics_14_cluster_feature_counts.pdf", width = 5, height = 3)
    print(p_hist)
  res <- dev.off()
} 

df <- multiomics_14_data_ann %>%
  group_by(Type, Cluster) %>%
  count() %>%
  ungroup()

p <- ggplot(df, aes(x = n, fill = Type)) +
  geom_density() +  
  scale_fill_manual(values = assay_cols) +
  labs(title = "multiomics_14 cluster analysis",
       x = "feature count",
       fill = "Assay") +
  facet_wrap(~Type, scales = "free") +
  theme_bw()+
  theme(#axis.title = element_blank(),
    #axis.text = element_blank(),
    axis.ticks = element_blank(),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank())
p

#Barplots of mean/median module activity for each ligand: please play around a bit to see if you can make compact version of these
# Thanks @Mark. Can you make these changes:
# Make bars narrower
# re-order into two rows: top) EGF, HGF, OSM; bottom) BMP, IFN, TGFB
# remove grid
# And then make two more iterations:
# Color bars by module (as in this posted version)
# Color bars by treatment using standard ligand colors
ligand_facet_order <- c( "EGF", "HGF", "OSM", "BMP2+EGF", "IFNG+EGF", "TGFB+EGF")
df <- multiomics_14_features_data_mean %>%
  pivot_longer(cols = contains("_")) %>%
  mutate(ligand = str_remove(name, "_.*"),
         ligand = factor(ligand, levels = ligand_facet_order)) %>%
  group_by(Cluster, ligand) %>%
  summarise(mean = mean(value), .groups = "drop")

p_cluster <- ggplot(df, aes(x = Cluster, y =mean, fill = Cluster))+
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = cluster14_cols) +
  theme(axis.text.x = element_text(angle = 90,
                                   size = 5),
        legend.key.size = unit(5,"points"),
        legend.text = element_text(size = 3),
        legend.title = element_text(size = 6),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    strip.background = element_rect(fill = "transparent")) +
  facet_wrap(~ligand)
p_cluster

p_ligand <- ggplot(df, aes(x = Cluster, y =mean, fill = ligand))+
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = ligand_cols) +
  # theme(axis.text.x = element_text(angle = 90,
  #                                  size = 5),
  #       legend.key.size = unit(5,"points"),
  #       legend.text = element_text(size = 5),
  #       legend.title = element_text(size = 6),
  #   panel.grid.major = element_blank(),
  #   panel.grid.minor = element_blank(),
  #   strip.background = element_rect(fill = "transparent")) +
  theme(axis.ticks = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines")) +
  facet_wrap(~ligand)
p_ligand

if(create_pdfs){
  pdf("../plots/MDD_multiomics_14_cluster_ligand_mean_bars.pdf", width = 4, height = 3)
  #print(p_cluster)
  print(p_ligand)
  res <- dev.off()
} 

#OSM-specific
df <- multiomics_14_features_data_mean %>%
  pivot_longer(cols = contains("_")) %>%
  filter(str_detect(name, "OSM_")) %>%
  mutate(ligand = str_remove(name, "_.*")) %>%
  group_by(Cluster, ligand) %>%
  summarise(mean = mean(value), .groups = "drop")

OSM_mean <- ggplot(df, aes(x = Cluster, y =mean, fill = ligand))+
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = ligand_cols) +
  scale_y_continuous(breaks = c(-1.5, -1, -.5, 0, .5, 1, 1.5),
                     labels = c("", -1, "", 0, "", 1, "")) +
  theme(axis.ticks.x = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x =element_blank(),
    axis.line.y = element_line(),
    axis.ticks.y = element_line(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines"))
OSM_mean

#Module 2 only
df <- multiomics_14_features_data_mean %>%
  pivot_longer(cols = contains("_")) %>%
  filter(Cluster == 2) %>%
  mutate(ligand = str_remove(name, "_.*")) %>%
  group_by(Cluster, ligand) %>%
  summarise(mean = mean(value), .groups = "drop")

cl2_mean <- ggplot(df, aes(x = ligand, y =mean, fill = ligand))+
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = ligand_cols) +
  scale_y_continuous(breaks = c(-1.5, -1, -.5, 0, .5, 1, 1.5),
                     labels = c("", -1, "", 0, "", 1, "")) +
  theme(axis.ticks.x = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x =element_blank(),
    axis.line.y = element_line(),
    axis.ticks.y = element_line(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines"))
cl2_mean

if(create_pdfs){
  pdf("../plots/MDD_multiomics_14_OSM_cl2.pdf", width = 4, height = 3)
  print(OSM_mean)
  print(cl2_mean)
  res <- dev.off()
} 

```

```{r RNAseq_influence}
#Birtwistle asked about how much of the module analysis was driven by the RNAseq data. I don’t want to spend a lot of time on this b/c I think it is adjacent to our goal of integrating over all data types, but I think there is one quick thing we can do: take the RNAseq data included in the megamatrix file to construct modules and cluster it. Then perform a confusion matrix analysis to compare overlap of RNAseq-only clusters to the distribution of RNAseq features for clusters you identified already. Assume that your integrative clusters represent ‘truth’. https://www.rdocumentation.org/packages/caret/versions/3.45/topics/confusionMatrix. If we get good overlap of clusters, we would conclude RNAseq is driving clusters; if we do not, then we would say that inclusion of other assay types is important for identification of integrative clusters. Thoughts on this?

if(file.exists(paste0("../integrated_analysis/MDD_RNAseq_clusters.rda"))){
    load(paste0("../integrated_analysis/MDD_RNAseq_clusters.rda"))
  } else {
    set.seed(42)
    combined18_RNAseq <- multiomics_dm[str_detect(rownames(multiomics_dm), ".*_RNAseq"),]
    medoids <- sample(1:nrow(combined18_RNAseq), size = multiomics_clust_num, replace = FALSE)
    clusters  <- pam(x = combined18_RNAseq, k = multiomics_clust_num, medoids = medoids, cluster.only=TRUE, pamonce = 5)
    combined18_RNAseq_data_ann <- as_tibble(combined18_RNAseq, rownames = "feature_type") %>%
      mutate(Cluster = clusters)

    combined18_RNAseq_clusters <- tibble(feature_type = rownames(combined18_RNAseq),
                                     Cluster = clusters)
    save(combined18_RNAseq_clusters,combined18_RNAseq_data_ann, file = paste0("../integrated_analysis/MDD_RNAseq_clusters.rda"))
  }

library(caret)


#find correlations between the modules and show cut to combine
# #Modules only set
# if_module_dm_reduced <- if_module_dm[,str_detect(colnames(if_module_dm),"Module")]

#Create a matrix and heatmap of the mean combined18 values
combined18_RNAseq_mean <- combined18_RNAseq_data_ann %>%
  select(-feature_type) %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

combined18_RNAseq_mean_dm <- combined18_RNAseq_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(combined18_RNAseq_mean_dm) <- paste0("Module_",combined18_RNAseq_mean$Cluster)

cor_dm <- cor(t(combined18_RNAseq_mean_dm),  method = "pearson")
hr = hclust(dist(cor_dm))
clusters = dendextend::cutree(hr, k = 14)
##Adding cuttree
##
cor_hm <- Heatmap(cor_dm,
                   name = "pearson\ncorrelation",
                  row_names_gp = gpar(fontsize = 10),
                  split = clusters,
                  column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_hm

multiomics_14_RNAseq_cluster_names <-c("A", "B", LETTERS[4:11], "C1","C2","C3","C4")

#Use the combined18 RNAseq data and the cluster assignments to
#create the multiomics_14 RNAseq dataset

multiomics_14_RNAseq_features_data_ann <- combined18_RNAseq_data_ann %>%
    mutate(Cluster = as.character(Cluster),
           Cluster = case_when(Cluster == "1" ~"A",
                               Cluster %in% c("2","17") ~"C1",
                               Cluster == "3" ~"B",
                               Cluster == "4" ~"D",
                               Cluster %in% c("5","13") ~"C2",
                               Cluster == "6" ~"E",
                               Cluster %in% c("7", "16") ~"C3",
                               Cluster == "8" ~"F",
                               Cluster == "9" ~"G",
                               Cluster %in% c("10","12") ~"C4",
                               Cluster == "11" ~"H",
                               Cluster == "14" ~"I",
                               Cluster == "15" ~"J",
                               Cluster == "18" ~"K"),
      Cluster = factor(Cluster, levels = multiomics_14_RNAseq_cluster_names, ordered = TRUE)) %>%
  as.data.frame()

#Remove the cluster assignments to keep the code similar to the 
#other dataset processing
# multiomics_14_RNAseq_features_data <- multiomics_14_RNAseq_features_data_ann %>%
#   select(-Cluster)

# #create a numeric matrix with feature row names
# multiomics_14_RNAseq_features_dm <- multiomics_14_RNAseq_features_data %>%
#   select(all_of(condition_order)) %>%
#   as.matrix() 
# rownames(multiomics_14_RNAseq_features_dm) <- multiomics_14_RNAseq_features_data$feature_type
# multiomics_14_RNAseq_features_dm <- multiomics_14_RNAseq_features_dm[!is.nan(multiomics_14_RNAseq_features_dm[,1]),]
# 
multiomics_14_RNAseq_clusters <- multiomics_14_RNAseq_features_data_ann %>%
  select(feature_type, Cluster)

df <- multiomics_14_data_ann %>%
  filter(Type == "RNAseq") %>%
  select(feature_type, Cluster) %>%
  rename(Integrated_cluster = Cluster) %>%
  mutate(Integrated_cluster = case_when(Integrated_cluster == "1" ~"A",
                                        Integrated_cluster == "2" ~"B",
                                        Integrated_cluster == "5" ~"D",
                                        Integrated_cluster == "7" ~"E",
                                        Integrated_cluster == "9" ~"F",
                                        Integrated_cluster == "11" ~"G",
                                        Integrated_cluster == "12" ~"H",
                                        Integrated_cluster == "13" ~"I",
                                        Integrated_cluster == "14" ~"J",
                                        Integrated_cluster == "18" ~"K",
                                        Integrated_cluster == "3+4" ~"C1",
                                        Integrated_cluster == "6+16" ~"C2",
                                        Integrated_cluster == "8+17" ~"C3",
                                        Integrated_cluster == "10+15" ~"C4"),
         Integrated_cluster = factor(Integrated_cluster, levels = multiomics_14_RNAseq_cluster_names, ordered = TRUE)) %>%
  left_join(multiomics_14_RNAseq_clusters, by = "feature_type") %>%
    rename(RNAseq_only_cluster = Cluster)

cf_matrix <- confusionMatrix(factor(df$RNAseq_only_cluster), df$Integrated_cluster)

#####Cluster the integrated matrix with and without non-RNAseq features 
#and with k = 14


if(file.exists(paste0("../integrated_analysis/MDD_14_clusters.rda"))){
    load(paste0("../integrated_analysis/MDD_14_clusters.rda"))
  } else {
    set.seed(42)
    medoids <- sample(1:nrow(multiomics_dm), size = 14, replace = FALSE)
    clusters  <- pam(x = multiomics_dm, k = 14, medoids = medoids, cluster.only=TRUE, pamonce = 5)
    im_14_data_ann <- as_tibble(multiomics_dm, rownames = "feature_type") %>%
      mutate(Cluster = clusters)
    
    combined18_RNAseq_dm <- multiomics_dm[str_detect(rownames(multiomics_dm), ".*_RNAseq"),]
    set.seed(42)
    medoids <- sample(1:nrow(combined18_RNAseq_dm), size = 14, replace = FALSE)
    clusters  <- pam(x = combined18_RNAseq_dm, k = 14, medoids = medoids, cluster.only=TRUE, pamonce = 5)
    im_14_RNAseq_data_ann <- as_tibble(combined18_RNAseq_dm, rownames = "feature_type") %>%
      mutate(Cluster = clusters)
    im_14_RNAseq_clusters <- tibble(feature_type = rownames(combined18_RNAseq_dm),
                                     Cluster = clusters)
    save(im_14_data_ann,
         im_14_RNAseq_data_ann,
         im_14_RNAseq_clusters,
         file = paste0("../integrated_analysis/MDD_14_clusters.rda"))
  }


df_14 <- im_14_data_ann %>%
  select(feature_type, Cluster) %>%
  filter(str_detect(feature_type, "RNAseq")) %>%
  rename(Integrated_cluster = Cluster) %>%
left_join(im_14_RNAseq_clusters, by = "feature_type") %>%
    rename(RNAseq_cluster = Cluster) %>%
  mutate(Integrated_cluster = as_factor(Integrated_cluster),
         RNAseq_cluster = as_factor(RNAseq_cluster))

cf_14_matrix <- confusionMatrix(df_14$RNAseq_cluster, reference = df_14$Integrated_cluster)

#14 Int clusters and 14 RNAseq clusters
df_int14 <- multiomics_14_data_ann %>%
 select(feature_type, Cluster) %>%
  filter(str_detect(feature_type, "RNAseq")) %>%
  rename(Integrated_cluster = Cluster) %>%
   mutate(Integrated_cluster = case_when(Integrated_cluster == "1" ~"1",
                                        Integrated_cluster == "2" ~"2",
                                        Integrated_cluster == "5" ~"5",
                                        Integrated_cluster == "7" ~"7",
                                        Integrated_cluster == "9" ~"9",
                                        Integrated_cluster == "11" ~"11",
                                        Integrated_cluster == "12" ~"12",
                                        Integrated_cluster == "13" ~"13",
                                        Integrated_cluster == "14" ~"14",
                                        Integrated_cluster == "18" ~"4",
                                        Integrated_cluster == "3+4" ~"3",
                                        Integrated_cluster == "6+16" ~"6",
                                        Integrated_cluster == "8+17" ~"8",
                                        Integrated_cluster == "10+15" ~"10")) %>%
left_join(im_14_RNAseq_clusters, by = "feature_type") %>%
    rename(RNAseq_cluster = Cluster) %>%
  mutate(Integrated_cluster = as_factor(Integrated_cluster),
         RNAseq_cluster = as_factor(RNAseq_cluster))

cf_int_14_matrix <- confusionMatrix(df_int14$RNAseq_cluster, reference = df_int14$Integrated_cluster)


#Compare stability of the Int 14 clusters that have different starting points

generate_ConfusionMatrix <- function(df, df_ref){
  df_int14 <- df_ref %>%
 select(feature_type, Cluster) %>%
  rename(Integrated_cluster = Cluster) %>%
   mutate(Integrated_cluster = case_when(Integrated_cluster == "1" ~"1",
                                        Integrated_cluster == "2" ~"2",
                                        Integrated_cluster == "5" ~"5",
                                        Integrated_cluster == "7" ~"7",
                                        Integrated_cluster == "9" ~"9",
                                        Integrated_cluster == "11" ~"11",
                                        Integrated_cluster == "12" ~"12",
                                        Integrated_cluster == "13" ~"13",
                                        Integrated_cluster == "14" ~"14",
                                        Integrated_cluster == "18" ~"4",
                                        Integrated_cluster == "3+4" ~"3",
                                        Integrated_cluster == "6+16" ~"6",
                                        Integrated_cluster == "8+17" ~"8",
                                        Integrated_cluster == "10+15" ~"10")) %>%
left_join(im_14_RNAseq_clusters, by = "feature_type") %>%
    rename(RNAseq_cluster = Cluster) %>%
  mutate(Integrated_cluster = as_factor(Integrated_cluster),
         RNAseq_cluster = as_factor(RNAseq_cluster))

}
  
  
```



This analysis can be interpreted to select the number of clusters in the dataset. It is based on a method described in Modern Statistics for Modern Biology by Susan Holmes and Wolfgang Huber http://web.stanford.edu/class/bios221/book/Chap-Clustering.html and is excerpted below.  

Taking the logarithm of the within-sum-of-squares (log(WSSk)) and comparing it to averages from simulated data with less structure can be a good way of choosing k. This is the basic idea of the gap statistic introduced by Tibshirani, Walther, and Hastie (2001). We compute log(WSSk) for a range of values of k, the number of clusters, and compare it to that obtained on reference data of similar dimensions with various possible ‘non-clustered’ distributions. We can use uniformly distributed data as we did above or data simulated with the same covariance structure as our original data.

The default choice for the number of clusters, k1, is the first value of k for which the gap is not larger than the first local maximum minus a standard error (see the manual page of the clusGap function). ... the choice recommended by Tibshirani, Walther, and Hastie (2001) is the smallest k such that gap(k)≥gap(k+1)−s′k+1...

#### combined18 dataset analysis  

Three of the methods for choosing k using a gap statitistic are:  
"Tibs2001SEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("Tibs2001SEmax"),
      SE.factor = 1) `  
      
uses the criterion, Tibshirani et al (2001) proposed: “the smallest k such that f(k) ≥ f(k+1) - s_{k+1}”. Note that this chooses k = 1 when all standard deviations are larger than the differences f(k+1) - f(k).


"firstSEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("firstSEmax"),
      SE.factor = 1) `  
      
location of the first f() value which is not smaller than the first local maximum minus SE.factor * SE.f[], i.e, within an “f S.E.” range of that maximum (see also SE.factor).

This, the default, has been proposed by Martin Maechler in 2012, when adding clusGap() to the cluster package, after having seen the "globalSEmax" proposal (in code) and read the "Tibs2001SEmax" proposal.

"globalSEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("globalSEmax"),
      SE.factor = 1)`    
      
(used in Dudoit and Fridlyand (2002), supposedly following Tibshirani's proposition): location of the first f() value which is not smaller than the global maximum minus SE.factor * SE.f[], i.e, within an “f S.E.” range of that maximum (see also SE.factor).



#### gss_RNAseq_combined18

Three of the methods for choosing k using a gap statitistic are:  
"Tibs2001SEmax": `r maxSE(gss_RNAseq_combined18$Tab[,'gap'], SE.f = gss_RNAseq_combined18$Tab[,'SE.sim'],
      method = c("Tibs2001SEmax"),
      SE.factor = 1) `  
      
uses the criterion, Tibshirani et al (2001) proposed: “the smallest k such that f(k) ≥ f(k+1) - s_{k+1}”. Note that this chooses k = 1 when all standard deviations are larger than the differences f(k+1) - f(k).


"firstSEmax": `r maxSE(gss_RNAseq_combined18$Tab[,'gap'], SE.f = gss_RNAseq_combined18$Tab[,'SE.sim'],
      method = c("firstSEmax"),
      SE.factor = 1) `  
      
location of the first f() value which is not smaller than the first local maximum minus SE.factor * SE.f[], i.e, within an “f S.E.” range of that maximum (see also SE.factor).

This, the default, has been proposed by Martin Maechler in 2012, when adding clusGap() to the cluster package, after having seen the "globalSEmax" proposal (in code) and read the "Tibs2001SEmax" proposal.

"globalSEmax": `r maxSE(gss_RNAseq_combined18$Tab[,'gap'], SE.f = gss_RNAseq_combined18$Tab[,'SE.sim'],
      method = c("globalSEmax"),
      SE.factor = 1)`    
###




```{r compare_to_previous, eval = FALSE}
### z score vs DE analysis
path <- "../integrated_analysis/Data/MDD_ligand_module_features.xlsx"
integrated_data <- lapply(readxl::excel_sheets(path), readxl::read_xlsx, path = path ) %>%
  bind_rows(.id = "module") %>%
  mutate(module = readxl::excel_sheets(path)[as.integer(module)])

int_genes <- integrated_data %>%
  filter(assay == "RNA",
         time %in% c(24,48)) %>%
  rename(Cluster = module) %>%
  select(Cluster, symbol) %>%
  mutate(Cluster = str_replace(Cluster, "module_","int_")) %>%
  distinct

set_genes <- multiomics_data_ann %>%
  filter(Type == "RNAseq") %>%
  rename(symbol = feature) %>%
  select(Cluster, symbol) %>%
  mutate(Cluster = paste0("set_",Cluster)) %>%
  distinct

genes <- bind_rows(int_genes, set_genes) %>%
  mutate(member = 1) %>%
  pivot_wider(names_from = Cluster, values_from = member, values_fill = 0) %>%
  rename(Name = symbol) %>%
  mutate(Name = as.factor(Name),
         across(where(is.numeric), as.integer)) %>%
  as.data.frame()

upset_obj <- upset(genes, nset = 32)

upset_obj

pdf("../plots/combined18_comparisons.pdf", width = 10, height = 8)
upset_obj
res = dev.off()
#######```
# We want to compare the features identified using set analysis of differentially expressed features to the prior method based on z scores of transformed assay values.  
# 
# #### Method  
# 
# Filter both matrices to 24 and 48 hour, RNAseq features and perform a set analysis on the modules in each dataset.  
```



### Compare T0 normalized

```{r compare_to_T0_normed, eval = TRUE}
#method -use original data assay_pk_data
#filter to assays and features in DE data
#T0 normalize
#use multiomics_14 cluster assignment in heatmap

#need to rename cycIF intensity features to match names in T0 Norm dataset
cluster_assignments <- multiomics_14_data_ann %>%
  select(Cluster, feature_type) %>%
  mutate(feature_type = case_when(str_detect(feature_type, "int_mean_") ~str_replace(feature_type,  "_.*int_mean_.*_cycIF", "_cycIF"),
                                  TRUE ~feature_type))

ligand_order_all_times <- paste(rep(ligand_order, each = 5), rep(c(1,4,8,24,48), times = 6), sep = "_")
  ligand_order_all_times <- factor(ligand_order_all_times, levels = ligand_order_all_times, ordered = TRUE)
  
#convert cycIF features to the average cell intensity values
cycIF_multiomics_14_features_int <- multiomics_14_data_ann %>%
  filter(Type == "cycIF",
         str_detect(feature, "_int_mean")) %>%
  mutate(feature = str_remove(feature, "_.*")) %>%
  group_by(feature) %>%
  summarise(across(where(is.numeric), mean), .groups = "drop") %>%
  mutate(feature_type = paste(feature, "cycIF", sep = "_")) %>%
  pivot_longer(cols = where(is.numeric), names_to = "condition")

#the T0 values don't map to the DE features so clusters can not be assigned
#some cyto vs nuc are in different clusers
multiomics_14_features_with_cycIF_int <- c(multiomics_14_data_ann$feature_type, cycIF_multiomics_14_features_int$feature_type)
  
#reformat raw assay metadata to match T0_DE
matched_features_types <- assay_pk_data %>%
    mutate(Type = case_when(Type == "RNA" ~"RNAseq",
                            TRUE ~ Type),
           ligand = case_when(ligand == "BMP2" ~"BMP2+EGF",
                              ligand == "IFNG" ~"IFNG+EGF",
                              ligand == "TGFB" ~"TGFB+EGF",
                              TRUE ~ ligand),
           feature = str_remove(feature, "_.*"),
           feature_type =  paste0(feature, "_", Type),
           condition = paste0(ligand,"_", experimentalTimePoint)) %>%
  select(-ligand, -feature, -Type, -experimentalTimePoint) %>%
  filter(feature_type %in% multiomics_14_features_with_cycIF_int)

assay_pk_data_T0_norm <- matched_features_types %>%
  filter(condition == "ctrl_0") %>%
  rename(ctrl_0_value = value) %>%
  select(-condition) %>%
  right_join(matched_features_types, by = "feature_type") %>%
  mutate(value_T0_norm = value-ctrl_0_value) %>%
  select(-ctrl_0_value, -value) %>%
pivot_wider(names_from = condition, values_from = value_T0_norm, values_fn = mean) %>%
  left_join(cluster_assignments, by = "feature_type") %>%
  select(feature_type, all_of(ligand_order_all_times), Cluster)

#create a numeric matrix with feature row names
assay_pk_data_T0_norm_dm <- assay_pk_data_T0_norm %>%
  select(-feature_type, -Cluster) %>%
  as.matrix() 
rownames(assay_pk_data_T0_norm_dm) <- assay_pk_data_T0_norm$feature_type

#Select only the 24 and 48 hour values
assay_pk_data_T0_norm_24_48 <- assay_pk_data_T0_norm %>%
  select(-matches("_1$|_4$|_8$"))

#create a numeric matrix with feature row names
assay_pk_data_T0_norm_24_48_dm <- assay_pk_data_T0_norm_24_48 %>%
  select(-feature_type, -Cluster) %>%
  as.matrix() 
rownames(assay_pk_data_T0_norm_24_48_dm) <- assay_pk_data_T0_norm_24_48$feature_type

#Calculate the mean values of each conditionxmodule set
assay_pk_data_T0_norm_mean <- assay_pk_data_T0_norm %>%
  pivot_longer(cols = matches("_[[:digit:]]"), names_to = "condition") %>%
  group_by(condition, Cluster) %>%
  summarise(value_mean =mean(value), .groups = "drop") %>%
  pivot_wider(names_from = condition, values_from = value_mean)

#Calculate the 24 and 48 mean values
assay_pk_data_T0_norm_24_48_mean <- assay_pk_data_T0_norm_mean %>%
  select(Cluster, matches("_24|_48"))
#create a numeric matrix with feature row names
assay_pk_data_T0_norm_24_48_mean_dm <- assay_pk_data_T0_norm_24_48_mean %>%
  select(-Cluster) %>%
  as.matrix() 
rownames(assay_pk_data_T0_norm_24_48_mean_dm) <- assay_pk_data_T0_norm_24_48_mean$Cluster

#prepare annotations to be used in heatmaps
assay_pk_data_T0_norm_ann <- assay_pk_data_T0_norm_24_48 %>%
  select(feature_type, Cluster) %>%
  mutate(Type = str_remove(feature_type, ".*_")) %>%
  select(-feature_type) %>%
  as.data.frame()
#######End of data prep

#heatmap of all T0 norm data
haRow_T0 <- HeatmapAnnotation(df = assay_pk_data_T0_norm_ann,
                              which = "row",
                              col = list(Cluster = factor(cluster14_cols, levels = cluster14_cols, ordered = TRUE),
                                         Type = assay_cols))
#heatmap of all T0 norm data
hm_T0 <- Heatmap(assay_pk_data_T0_norm_dm,
        name = "fc ",
        column_title = "T0 Normalized, multiomics_14 structure",
        column_title_gp = gpar(fontsize = 12),
        col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
        cluster_rows = FALSE,
        cluster_row_slices = FALSE,
        cluster_columns = FALSE,
        row_split = assay_pk_data_T0_norm$Cluster,
        #row_gap = unit(2, "mm"),
        row_title = " ",
        left_annotation = haRow_T0,
        show_row_names = FALSE,
        row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 10),
        use_raster = FALSE)

hm_T0
#heatmap of only 24 and 48 hour values
#heatmap of all T0 norm data
hm_T0_24_48 <- Heatmap(assay_pk_data_T0_norm_24_48_dm,
        name = "fc ",
        column_title = "T0 Normalized, multiomics_14 structure, 24, 48 hour only",
        column_title_gp = gpar(fontsize = 12),
        col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")),
        cluster_rows = FALSE,
        cluster_row_slices = FALSE,
        cluster_columns = FALSE,
        row_split = assay_pk_data_T0_norm$Cluster,
        #row_gap = unit(2, "mm"),
        row_title = " ",
        left_annotation = haRow_T0,
        show_row_names = FALSE,
        row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 10),
        use_raster = FALSE)
hm_T0_24_48

#heatmap of mean values
hm_T0_24_48_mean_cl <- Heatmap(assay_pk_data_T0_norm_24_48_mean_dm,
        name = "value ",
        column_title = "T0 Normalized means, multiomics_14 structure, clustered",
        column_title_gp = gpar(fontsize = 12),
        col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
        cluster_rows = TRUE,
        cluster_row_slices = FALSE,
        cluster_columns = FALSE,
        #row_gap = unit(2, "mm"),
        row_title = " ",
        show_row_names = TRUE,
        row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 10),
        use_raster = FALSE)
hm_T0_24_48_mean_cl

hm_T0_24_48_mean_or <- Heatmap(assay_pk_data_T0_norm_24_48_mean_dm,
        name = "value ",
        column_title = "T0 Normalized means, multiomics_14 structure, ordered",
        column_title_gp = gpar(fontsize = 12),
        col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
        cluster_rows = FALSE,
        cluster_row_slices = FALSE,
        cluster_columns = FALSE,
        #row_gap = unit(2, "mm"),
        row_title = " ",
        show_row_names = TRUE,
        row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 10),
        use_raster = FALSE)
hm_T0_24_48_mean_or


#barplots with clusters on x axis, mean values on y axis, faceted by condition
df <- assay_pk_data_T0_norm_24_48_mean %>%
  pivot_longer(cols = matches("_[[:digit:]]"), names_to = "condition") %>%
  mutate(condition = factor(condition, levels = ligand_order_all_times, ordered = TRUE))

mean_bar_plot <- ggplot(df, aes(x = Cluster, y = value)) +
  geom_col() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  facet_wrap(~condition, ncol = 2)
mean_bar_plot

if(create_pdfs){
  pdf("../plots/MDD_T0_normed_in_multiomics_14_clusters.pdf",height = 10)
  draw(hm_T0)
  draw(hm_T0_24_48)
  draw(hm_T0_24_48_mean_cl)
  draw(hm_T0_24_48_mean_or)
  print(mean_bar_plot)
  res <- dev.off()
} 

module_features <- assay_pk_data_T0_norm_24_48 %>%
  arrange(Cluster) %>%
  mutate(Cluster = paste0("module_", Cluster),
         Cluster = factor(Cluster, levels = unique(Cluster))) %>%
  split(.$Cluster) 

res <- write_xlsx(module_features,
                  path = "../tables/MDD_T0_norm_combined18_tables.xlsx")

#Cluster the T0 data according to the multiomics_14 structure
#and drop RNAseq
# assay_pk_data_T0_norm14 <- assay_pk_data_T0_norm %>%
#   select(-Cluster) %>%
#   left_join(cluster_assignments, by = "feature_type") %>%
#   filter(!str_detect(feature_type, "RNAseq")) %>%
#   select(all_of(ligand_order_all_times),feature_type, Cluster)
# 
# #create heatmap of T0 data using DE cluster assignments
# #create a numeric matrix with feature row names
# assay_pk_data_T0_norm14_dm <- assay_pk_data_T0_norm14 %>%
#   select(-feature_type, -Cluster) %>%
#   as.matrix() %>%
#   t %>%
#   scale %>%
#   t
# rownames(assay_pk_data_T0_norm14_dm) <- assay_pk_data_T0_norm14$feature_type
# 
# assay_pk_data_T0_norm14_ann <- assay_pk_data_T0_norm14 %>%
#   select(feature_type, Cluster) %>%
#   mutate(Type = str_remove(feature_type, ".*_")) %>%
#   select(-feature_type) %>%
#   as.data.frame()
# 
# haRow_T0_14 <- HeatmapAnnotation(df = assay_pk_data_T0_norm14_ann,
#                               which = "row",
#                               col = list(Cluster = factor(cluster14_cols, levels = cluster14_cols, ordered = TRUE),
#                                          Type = assay_cols))
# 
# hm_T0_14 <- Heatmap(assay_pk_data_T0_norm14_dm,
#         name = "fc ",
#         column_title = "T0 Normalized, multiomics_14 structure",
#         column_title_gp = gpar(fontsize = 12),
#         col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")),
#         cluster_rows = FALSE,
#         cluster_row_slices = FALSE,
#         cluster_columns = FALSE,
#         row_split = assay_pk_data_T0_norm14$Cluster,
#         #row_gap = unit(2, "mm"),
#         row_title = " ",
#         left_annotation = haRow_T0_14,
#         show_row_names = TRUE,
#         row_names_gp = gpar(fontsize = 4),
#         column_names_gp = gpar(fontsize = 10),
#         use_raster = FALSE)
# 
# hm_T0_14
# 
# #Make a heatmap of the mean values
# all_conditions_without_1 <- str_remove_all(ligand_order_all_times, ".*_1") %>% stringi::stri_remove_empty()
# 
# assay_pk_data_T0_norm14_mean <- assay_pk_data_T0_norm14 %>%
#   filter(!str_detect(feature_type, "RNAseq|motifs")) %>%
#   pivot_longer(cols = where(is.numeric), names_to = "condition") %>%
#   group_by(condition, Cluster) %>%
#   summarise(value_mean =mean(value), .groups = "drop") %>%
#   pivot_wider(names_from = condition, values_from = value_mean) %>%
#   select(all_of(all_conditions_without_1), Cluster)
#   
# #create heatmap of T0 data using DE cluster assignments
# #create a numeric matrix with feature row names
# assay_pk_data_T0_norm14_mean_dm <- assay_pk_data_T0_norm14_mean %>%
#   select(-Cluster) %>%
#   as.matrix()  %>%
#   t %>%
#   scale %>%
#   t
# 
# rownames(assay_pk_data_T0_norm14_mean_dm) <- assay_pk_data_T0_norm14_mean$Cluster
# 
# hm_T0_norm14_mean <- Heatmap(assay_pk_data_T0_norm14_mean_dm,
#         name = "value ",
#         column_title = "T0 Normalized means, multiomics_14 structure",
#         column_title_gp = gpar(fontsize = 12),
#         col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")),
#         cluster_rows = TRUE,
#         cluster_row_slices = FALSE,
#         cluster_columns = FALSE,
#         #row_gap = unit(2, "mm"),
#         row_title = " ",
#         show_row_names = TRUE,
#         row_names_gp = gpar(fontsize = 8),
#         column_names_gp = gpar(fontsize = 10),
#         use_raster = FALSE)

# #barplots with clusters on x axis, mean values on y axis, facetted by condition
# condition_all_order <- paste0(rep(c("HGF", "OSM", "EGF", "BMP2+EGF", "IFNG+EGF", "TGFB+EGF"), each=4), rep(c("_4", "_8", "_24", "_48"), times = 6))
# 
# df <- assay_pk_data_T0_norm14_mean %>%
#   pivot_longer(cols = where(is.numeric), names_to = "condition") %>%
#   drop_na() %>%
#   mutate(condition = factor(condition, levels = condition_all_order, ordered = TRUE))
# 
# p <- ggplot(df, aes(x = Cluster, y = value)) +
#   geom_col() +
#   labs(title = "T0 normalized in multiomics_14 structure",
#        subtitle = "mean values per condition and module") +
#   theme_bw() +
#   theme(axis.text.x = element_text(angle = 90),
#         strip.background.x = element_rect(fill = "transparent")) +
#   facet_wrap(~condition, ncol = 4)
# p
# 
# if(create_pdfs){
#   pdf("../plots/MDD_T0_normed_multiomics_14.pdf",height = 10)
#   draw(hm_T0_14)
#   draw(hm_T0_norm14_mean)
#   print(p)
#   res <- dev.off()
# } 
# 
# 
# 
# if(create_pdfs){
#   pdf("../plots/MDD_T0_normed_in_combined18_manual_clusters.pdf",height = 10)
#   p
#   draw(hm_T0)
#   res <- dev.off()
# } 


direction_df <- unique_features_data_ann %>%
  select(feature_type, Direction, Ligand, Type) %>%
  filter(str_detect(feature_type, "_cycIF|_GCP|_RPPA"),
         !Ligand == "PBS")

cycIF_directions <- direction_df %>%
  filter(str_detect(feature_type, "int"),
         !str_detect(feature_type, "nfkbp65|p21waf1cip1_3_af647cy5_int_mean_nuc_")) %>%
  mutate(biomarker = str_remove(feature_type, "_.*"),
         feature_type = paste0(biomarker, "_cycIF")) %>%
  select(feature_type, Direction, Ligand, Type)

corrected_directions <- direction_df %>%
  filter(!str_detect(feature_type, "cycIF")) %>%
  bind_rows(cycIF_directions)

cycIF_GCP_RPPA_raw <- assay_pk_data %>%
  filter(Type %in% c("cycIF", "GCP", "RPPA"),
         !ligand == "PBS") %>%
    mutate(ligand = case_when(ligand == "BMP2" ~"BMP2+EGF",
                              ligand == "IFNG" ~"IFNG+EGF",
                              ligand == "TGFB" ~"TGFB+EGF",
                              TRUE ~ ligand),
           feature_type = feature,
           feature = str_remove(feature, "_RPPA|_cycIF|_GCP"),
           condition = paste0(ligand,"_", experimentalTimePoint)) %>%
  select(-ligand, -feature, -Type, -experimentalTimePoint)

cycIF_GCP_RPPA_T0_norm <- cycIF_GCP_RPPA_raw %>%
  filter(condition == "ctrl_0") %>%
  rename(ctrl_0_value = value) %>%
  select(-condition) %>%
  right_join(cycIF_GCP_RPPA_raw, by = "feature_type") %>%
  mutate(value_T0_norm = value-ctrl_0_value) %>%
  select(-ctrl_0_value, -value) %>%
pivot_wider(names_from = condition, values_from = value_T0_norm, values_fn = mean) %>%
  right_join(corrected_directions, by = "feature_type")  %>%
arrange(Ligand, Direction, Type) %>%
  select(ctrl_0, everything())
###DEBUG
#### need to reconcile that assay_pk_data cycIF contains the average intensity values while the T0_DE contains individual cycIF values.
# 
# df <- unique_features_data_ann %>%
#   filter(!Type %in% c("RNAseq", "motifs")) %>%
#   arrange(Set, Direction)
# 
df_dm<- cycIF_GCP_RPPA_T0_norm %>%
    select(where(is.numeric)) %>%
  as.matrix() %>%
  t() %>%
  scale() %>%
  t()
rownames(df_dm) <- cycIF_GCP_RPPA_T0_norm$feature_type

df_ann <- cycIF_GCP_RPPA_T0_norm %>%
    select(Ligand, Direction, Type) %>%
  as.data.frame()

haRow_df <- HeatmapAnnotation(df = df_ann,
                                  which = "row",
                                  col = list(Ligand = factor(ligand_cols, levels = ligand_cols, ordered = TRUE),
                                             Type = assay_cols,
                                             Direction = factor(c("Negative" = "blue","Mixed" = "gray","Positive" = "red"))))
# 
hm_T0 <- Heatmap(df_dm,
               name = "lfc rr",
               column_title = "Ligand-specific, T0 norm of cycIF, GCP, RPPA",
               column_title_gp = gpar(fontsize = 12),
               col = colorRamp2(c(-2, 0, 2), c("#2166AC", "white", "#B2182B")),
               cluster_rows = FALSE,
               cluster_row_slices = FALSE,
               cluster_columns = FALSE,
               #row_split = df_ann$Cluster,
               row_gap = unit(2, "mm"),
               row_title = " ",
               show_row_names = TRUE,
               row_names_gp = gpar(fontsize = 7),
               column_names_gp = gpar(fontsize = 7),
               left_annotation = haRow_df,
               use_raster = FALSE)
hm_T0

if(create_pdfs){
  pdf("../plots/MDD_integrated_T0_cycIF_GCP_RPPA_heatmap.pdf")
  draw(hm_T0)
  res <- dev.off()
}


#Add to Compare T0 Normalized tab an analysis of multiomics_14 modules and drop CTRL. Also simplify bar plots to remove gray background (Mark)

```

##### methods  
load original data  
filter to assays and features in DE data  
T0 normalize  
use combined18 cluster assignment in heatmap  

##### T0 normalized in multiomics_14 structure  
load original data  
filter to features in the multiomics_14 dataset  
retain 4,8, 24 and 48 hour data  
center on ctrl_0 values  
calculate mean of each condition for each of the 14 modules  
display as facetted bar graph  

###

### Correlations

```{r correlations}

#Ligand cross-correlation heat map with final megamatrix feature list: 24H only, 48H only, 24+48H combined. Explore scaling options. (Mark)

all_data_cor <- multiomics_data %>%
  select(matches("_24|_48")) %>%
  as.matrix() %>%
  cor(method = "pearson")

all_data_hm <- Heatmap(all_data_cor,
                       column_title = "All data, T0 DE correlations",
                   name = "pearson\ncorrelation ",
                  row_names_gp = gpar(fontsize = 10),
                  column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
all_data_hm

if(write_csvs){
  res <- write_csv(as_tibble(all_data_cor, rownames = "Condition")%>%
  mutate(across(where(is.numeric), signif, digits = 3)),
  path = "../tables/MDD_integrated_condition_condition_correlations.csv")
}

all_data_24_cor <- multiomics_data %>%
  select(matches("_24")) %>%
  as.matrix() %>%
  cor(method = "pearson")

cor_24_hm <- Heatmap(all_data_24_cor,
                     column_title = "24 hour, T0 DE correlations",
                     name = "pearson\ncorrelation ",
                     row_names_gp = gpar(fontsize = 10),
                     column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_24_hm

all_data_48_cor <- multiomics_data %>%
  select(matches("_48")) %>%
  as.matrix() %>%
  cor(method = "pearson")

cor_48_hm <- Heatmap(all_data_48_cor,
                     column_title = "48 hour, T0 DE correlations",
                     name = "pearson\ncorrelation ",
                     row_names_gp = gpar(fontsize = 10),
                     column_names_gp = gpar(fontsize = 10),
                     col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_48_hm


all_data_pk_cor <- assay_pk_data_T0_norm %>%
  select(matches("_24|_48")) %>%
  as.matrix() %>%
  cor(method = "pearson")

all_data_pk_hm <- Heatmap(all_data_pk_cor,
                          column_title = "All data, T0 normalized correlations",
                          name = "pearson\ncorrelation ",
                          row_names_gp = gpar(fontsize = 10),
                          column_names_gp = gpar(fontsize = 10),
                          col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
all_data_pk_hm

all_data_pk_24_cor <- assay_pk_data_T0_norm %>%
  select(matches("_24")) %>%
  as.matrix() %>%
  cor(method = "pearson")

cor_24_pk_hm <- Heatmap(all_data_pk_24_cor,
                        column_title = "24 hour data, T0 normalized correlations",
                        name = "pearson\ncorrelation ",
                        row_names_gp = gpar(fontsize = 10),
                        column_names_gp = gpar(fontsize = 10),
                        col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_24_pk_hm

all_data_pk_48_cor <- assay_pk_data_T0_norm %>%
  select(matches("_48")) %>%
  as.matrix() %>%
  cor(method = "pearson")

cor_48_pk_hm <- Heatmap(all_data_pk_48_cor,
                   column_title = "48 hour data, T0 normalized correlations",
                   name = "pearson\ncorrelation ",
                  row_names_gp = gpar(fontsize = 10),
                  column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_48_pk_hm

if(create_pdfs){
  pdf("../plots/MDD_integrated_cors_condition_condition.pdf", width = 7, height = 6) 
  draw(all_data_hm)
  res <- dev.off()
} 

if(create_pdfs){
  pdf("../plots/MDD_integrated_cors.pdf") 
  draw(all_data_hm)
  draw(cor_24_hm)
  draw(cor_48_hm)
  draw(all_data_pk_hm)
  draw(cor_24_pk_hm)
  draw(cor_48_pk_hm)
  res <- dev.off()
} 

```

###