---
title: "MDD integrated analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, cache=FALSE, cache.lazy = FALSE, message=FALSE, warning = FALSE, fig.height=10, fig.width=14)

suppressMessages(library(tidyverse))
library(ComplexHeatmap)
library(circlize)
library(umap)
library(cluster)
library(RColorBrewer)
library(writexl)
library(clue)
#library(UpSetR)

source("../R/MDD_functions.R")
create_pdfs <- TRUE
write_csvs <- TRUE

```

```{r functions}

create_consensus_clustering <- function(mat = multiomics_data_dm,
                                        nbr_clusterings = 5,
                                        nbr_consensus = 3,
                                        clust_num = multiomics_clust_num){
  cons_list <- lapply(seq_len(nbr_consensus), function(c) {
    cons <- cl_ensemble(list = lapply(seq_len(nbr_clusterings), function(b) { #Create clusterings
      medoids <- sample(nrow(mat), size = clust_num, replace = FALSE)
      clusters  <- pam(x = mat, 
                       k = clust_num, 
                       medoids = medoids, 
                       pamonce = 5)})) %>%
      cl_consensus(method = "HE", control = list(nruns = 5))
  }) %>%
    cl_consensus(method = "HE", control = list(nruns = 5))
}

```

```{r setup_parameters_and_metadata}
data_path <- "../multiomics_analysis/data/"

multiomics_all_features_matrix_path <- paste0(data_path, "MDD_multiomics_matrix.csv")
shared_features_path <- paste0(data_path,"MDD_multiomics_shared_features.csv") 
unique_features_path <- paste0(data_path,"MDD_multiomics_unique_features.csv") 
multiomics_clust_num <- 18
cluster_method <- "pam"

experimentalTimePoints <- c("_24", "_48")
ligand_order <-  c("HGF", "OSM", "EGF","BMP2+EGF", "IFNG+EGF", "TGFB+EGF")

condition_order <- paste0(rep(ligand_order, each = length(experimentalTimePoints)), rep(c("_24", "_48"), times = length(ligand_order)))

#create a named vector of color values for the Heatmap clusters
cluster_cols <- c(structure(RColorBrewer::brewer.pal(12, "Paired")[c(1:12)], names = 1:12), c("13" = "#222222", "14" = "#555555", "15" = "#999999", "16" = "#AAAAAA", "17" = "#BBBBBB", "18" = "#CCCCCC", "19" = "#DDDDDD"))

ligand_cols <- c("CTRL" = "#7A4A2A",
                 "PBS" = "#8dd3c7",
                 "HGF" = "#80b1d3",
                 "OSM" = "#fdb462",
                 "EGF" = "#fb8072",
                 "BMP2+EGF" = "#b3de69",
                 "IFNG+EGF" = "#bebada",
                 "TGFB+EGF" = "#ffd92f")

```

The cluster method is `r cluster_method`.  
Input files are `r multiomics_all_features_matrix_path`, `r shared_features_path` and `r unique_features_path`.  

Method summary:

Use replicates to perform differential (DE) analysis between conditions and ctrl_0 for RNAseq, RPPA, GCP and cycIF assays. Use family z scores for motifs.  
rrscale the DE values within each assay  
Use set analysis to select features that are significantly expressed by a single ligand (ligand-specific) or multiple ligands (shared)  
Perform Gap analysis with pam clustering on the shared and ligand-specific values as a single dataset  
Combine 4 similar cluster/module pairs to yield a multiomics_14 dataset  Evaluate the multiomics_14 dataset with UMAP  
Correlate the multiomics_14 modules with IF and cell tracking phenotypes  
Get enriched Reactome pathways for the RNAseq genes each multiomics_14 module  
Evaluate the modules/assay content and distributions  
Evaluate T0 centered versions of the assay values using the multiomics_14 structure  

```{r prepData}

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(multiomics_all_features_matrix_path)) stop("cannot read ",multiomics_all_features_matrix_path)
all_features_data <- read_csv(multiomics_all_features_matrix_path) %>%
  pivot_longer(cols = matches("_24|_48"), names_to = "condition") %>%
  mutate(condition = str_replace(condition, "BMP2","BMP2+EGF"),
         condition = str_replace(condition, "IFNG","IFNG+EGF"),
         condition = str_replace(condition, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_")) %>%
  pivot_wider(names_from = condition, values_from = value) %>%
  select(feature, Type, feature_type, all_of(condition_order)) %>%
  drop_na

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(unique_features_path)) stop("cannot read ",unique_features_path)
set.seed(42) #randomize row positions
unique_features <- read_csv(unique_features_path) %>%
  slice_sample(prop = 1)
unique_features_data <- unique_features%>%
  mutate(ligand = str_replace(ligand, "BMP2","BMP2+EGF"),
         ligand = str_replace(ligand, "IFNG","IFNG+EGF"),
         ligand = str_replace(ligand, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_"),
         Set = ligand) %>%
  select(ligand, feature_type, Abs_Direction, Set) %>%
  rename(Direction = Abs_Direction) %>%
  left_join(all_features_data, by = "feature_type") %>%
  rename(Cluster = ligand) %>%
  mutate(Cluster = paste(Cluster, Direction, sep = "_")) %>%
  drop_na

#Read in the integrated data, add EGF to some names, set order
if(!file.exists(shared_features_path)) stop("cannot read ",shared_features_path)
set.seed(42) #randomize row positions
shared_features <- read_csv(shared_features_path) %>%
  slice_sample(prop = 1)
shared_features_data <- shared_features %>%
  mutate(ligand = str_replace(ligand, "BMP2","BMP2+EGF"),
         ligand = str_replace(ligand, "IFNG","IFNG+EGF"),
         ligand = str_replace(ligand, "TGFB","TGFB+EGF"),
         feature_type =  paste(feature, Type, sep = "_"),
         Set = "shared") %>%
  select(feature_type, Set) %>%
  distinct() %>%
  left_join(all_features_data, by = c("feature_type" ))  %>%
  drop_na

set.seed(42) #randomize row positions
multiomics_data <- bind_rows(shared_features_data, select(unique_features_data, -Direction, -Cluster)) %>%
  slice_sample(prop = 1)
assay_cols <- structure(RColorBrewer::brewer.pal(8, "Paired")[c(3, 2,6,5, 1)], names = unique(multiomics_data$Type))

```


```{r cluster_data}

#create a numeric matrix with feature row names
multiomics_data_dm <- multiomics_data %>%
  select(all_of(condition_order)) %>%
  as.matrix() 
rownames(multiomics_data_dm) <- multiomics_data$feature_type
multiomics_data_dm <- multiomics_data_dm[!is.nan(multiomics_data_dm[,1]),]

if(file.exists(paste0(data_path,"MDD_multiomics_clusters.csv"))){
  cluster_df <- read_csv(paste0(data_path,"MDD_multiomics_clusters.csv")) 
  clusters <- cluster_df %>%
    pull(Cluster)
  names(clusters) <- cluster_df %>%
    pull(feature_type)
} else {
  #     Consensus clusterings “synthesize” the information in the elements of a cluster ensemble into a single clustering, often by minimizing a criterion function measuring how dissimilar consensus candidates are from the (elements of) the ensemble (the so-called “optimization approach” to consensus clustering).
  # 
  # The most popular criterion functions are of the form L(x) = ∑ w_b d(x_b, x)^p, where d is a suitable dissimilarity measure (see cl_dissimilarity), w_b is the case weight given to element x_b of the ensemble, and p ≥ 1. If p = 1 and minimization is over all possible base clusterings, a consensus solution is called a median of the ensemble; if minimization is restricted to the elements of the ensemble, a consensus solution is called a medoid (see cl_medoid). For p = 2, we obtain least squares consensus partitions and hierarchies (generalized means). See also Gordon (1999) for more information.
  # 
  # If all elements of the ensemble are partitions, the built-in consensus methods compute consensus partitions by minimizing a criterion of the form L(x) = ∑ w_b d(x_b, x)^p over all hard or soft partitions x with a given (maximal) number k of classes. Available built-in methods are as follows.
  nbr_clusterings <- 100
  nbr_consensus <- 3
  nbr_runs <- 25

  set.seed(42)
  res <- lapply(1:nbr_runs, function(x, nbr_clusterings = nbr_clusterings, nbr_consensus = nbr_consensus){
    cons <- create_consensus_clustering(nbr_clusterings = nbr_clusterings, nbr_consensus= nbr_consensus) 
    save(cons, file =paste0("../Data/MDD_multiomics_clstrng_run_",x,".rda"))
  })
  con_cons <- clue::cl_ensemble(list = lapply(1:25, function(x){
    load(paste0("../Data/MDD_multiomics_clstrng_run_",x,".rda"))
    clue::as.cl_partition(cons)
  }))
 
  #generate a consensus of all clustering consensuses
  set.seed(100)
  ce_cons <- clue::cl_consensus(con_cons, method = "HE", control = list(nruns = 5))
  #Save the cluster assignments to disk
  clusters <- cl_class_ids(ce_cons)
  cluster_df <- tibble(Cluster = clusters,
                       feature_type = names(clusters))
  write_csv(cluster_df, paste0(data_path,"MDD_multiomics_clusters.csv")) 
  
  #Calculate the agreement between ensemble elements and the consensus clustering
  #
  #clue::cl_agreement()
  #"FM"  the index of Fowlkes and Mallows (1983), i.e., the ratio N_xy / sqrt(N_x N_y) of the number N_xy of distinct pairs of objects in the same class in both partitions and the geometric mean of the numbers N_x and N_y of distinct pairs of objects in the same class in partition x and partition y, respectively. For soft partitions, (currently) the Fowlkes-Mallows index of the corresponding nearest hard partitions is used.
  
   #calculate agreement between all possible cluster pairings
  agrmnts <- lapply(1:(nbr_runs-1), function(x){
    res <- lapply((x+1):nbr_runs, function(y){
      cl_agreement(con_cons[[x]], con_cons[[y]], method = "FM")
    }) %>%
      unlist()
  })
  
  p_inner_agreements <- ggplot(tibble(agreements = unlist(agrmnts)), aes(x = agreements)) +
    geom_density() +
    theme_bw()
  p_inner_agreements
  
  ags_con <- sapply(con_cons, clue::cl_agreement, y = ce_cons, method = "FM")
    p_agreement_to_consensus <- ggplot(tibble(agreement_to_consensus = ags_con, method = "FM"), aes(x = agreement_to_consensus)) +
    geom_density() +
    theme_bw()
  p_agreement_to_consensus
  
  #Todo annotate figure with statistics
  agreement_to_consensus_mean <- mean(ags_con)
  agreement_to_consensus_sd <- sd(ags_con)

}

#Add the cluster assignments to the data
multiomics_data_ann <-  tibble(feature_type = names(clusters),
                               Cluster = as.character(clusters)) %>%
  right_join(multiomics_data, by = "feature_type") %>%
  mutate(Cluster = factor(Cluster, ordered = TRUE))

multiomics_ann <- multiomics_data_ann %>%
  select(Cluster, Type, Set) %>%
  as.data.frame()

#Create a matrix and heatmap of the mean multiomics values
multiomics_data_mean <- multiomics_data_ann %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

multiomics_count <- multiomics_data_ann %>%
  dplyr::count(Cluster) 

multiomics_data_mean_dm <- multiomics_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_data_mean_dm) <- paste0("Module_",multiomics_data_mean$Cluster,"_", multiomics_count$n)

```

```{r correlate_modules}

#find correlations between the modules and show cut to combine into 14 modules
cor_dm <- cor(t(multiomics_data_mean_dm),  method = "pearson")
hr = hclust(dist(cor_dm))
clusters = dendextend::cutree(hr, k = 14)

cor_hm <- Heatmap(cor_dm,
                  column_title = "Module correlations",
                   name = "pearson\ncorrelation",
                  row_title = " ",
                  row_names_gp = gpar(fontsize = 10),
                  split = clusters,
                  column_names_gp = gpar(fontsize = 10),
                  col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
cor_hm

if(create_pdfs) {
  pdf("../multiomics_analysis/plots/MDD_SF6b.pdf",
      width = 9, height = 8)
  res <- print(cor_hm)
  res <- dev.off()
}

# if(write_csvs) {
#   write.csv(cor_dm, "../tables/MDD_multiomics_module_module_correlations.csv")
# }

```


```{r reduce_to_14_clusters}
#####
#Use the multiomics features data and the cluster assignments to
#create the multiomics_14 dataset

multiomics_14_clusters <-c("1", "2+8", "3", "4", "5+6", "7", "9+18", "10", "11","12","13","14+16", "15","17")
cluster14_cols <- c(structure(RColorBrewer::brewer.pal(12, "Paired")[1:12], names = multiomics_14_clusters[1:12]), c("15" = "#222222", "17" = "#555555"))
                          
multiomics_14_data_ann <- multiomics_data_ann %>%
    mutate(Cluster = as.character(Cluster),
      Cluster = case_when(Cluster %in% c("2","8") ~"2+8",
                          Cluster %in% c("5","6") ~"5+6",
                          Cluster %in% c("9", "18") ~"9+18",
                          Cluster %in% c("14","16") ~"14+16",
                               TRUE ~Cluster),
      Cluster = factor(Cluster, levels = multiomics_14_clusters, ordered = TRUE)) %>%
  as.data.frame()

#create a numeric matrix with feature row names
multiomics_14_dm <- multiomics_14_data_ann %>%
  select(all_of(condition_order)) %>%
  as.matrix() 
rownames(multiomics_14_dm) <- multiomics_14_data_ann$feature_type
multiomics_14_dm <- multiomics_14_dm[!is.nan(multiomics_14_dm[,1]),]

# 
# multiomics_14_features_count <- multiomics_14_data_ann %>%
#   group_by(Cluster) %>%
#   dplyr::count()

#Create a matrix and heatmap of the mean combined18 values
multiomics_14_data_mean <- multiomics_14_data_ann %>%
  group_by(Cluster) %>%
  summarise(across(.cols = matches("_24|48"), .fns = mean, .groups = "drop"))

multiomics_14_data_mean_dm <- multiomics_14_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_14_data_mean_dm) <-multiomics_14_data_mean$Cluster

```

```{r 14_module_heatmaps}

###multiomics_14 heatmap with ligand column annotations, RPPA, cycIF and motif callouts
  callouts <- read_csv(paste0(data_path, "MDD_multiomics_heatmap_callouts.csv")) %>%
    filter(selected == "x")

#Annotate callouts with abbreviations of the type 

df <- multiomics_14_data_ann %>%
  #filter(!Type == "RNAseq") %>%
  mutate(feature = str_remove(feature, "_[[:digit:]]_.*[nuc|cytoplasm]"),
         feature = case_when(Type == "RPPA" ~paste(feature, "(R)"),
                             Type == "cycIF" ~paste(feature, "(C)"),
                             Type == "motifs" ~paste(feature, "(A)"),
                             TRUE ~ feature))

RPPA_motif_rows_right <- which(df$feature_type %in% callouts$feature_type)

haRight = rowAnnotation(foo = anno_mark(at = RPPA_motif_rows_right,
                                   labels = df$feature[RPPA_motif_rows_right],
                                   labels_gp = gpar(fontsize = 5)))

hm_multiomics_14_callouts <- Heatmap(multiomics_14_dm,
                       name = "lfc rr",
                       column_title = "multiomics_14 features",
                       column_title_gp = gpar(fontsize = 6),
                       width = ncol(multiomics_14_dm)*unit(4, "mm"),
                       col = colorRamp2(c(-1.5, 0, 1.5), c("#2166AC", "white", "#B2182B")),
                       cluster_rows = FALSE,
                       cluster_row_slices = FALSE, 
                       cluster_columns = FALSE,
                       row_split = multiomics_14_data_ann$Cluster,
                       row_gap = unit(3, "mm"),
                       #row_title = " ",
                       show_row_names = FALSE,
                       row_names_gp = gpar(fontsize = 12),
                       column_names_gp = gpar(fontsize = 6),
                     # left_annotation = haLeft,
                       right_annotation = haRight,
                       #top_annotation = haCol_multiomics_14,
                       use_raster = FALSE)
hm_multiomics_14_callouts


if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_F5a.pdf",height = 10) 
  draw(hm_multiomics_14_callouts)
  res <- dev.off()
} 

#Compare the consensus clustering to the previous single clustering
path <- "../multiomics_analysis/data/MDD_combined14_single_tables.xlsx"
multiomics_combined14_single <- path %>% 
  readxl::excel_sheets() %>% 
  set_names() %>% 
  map_dfr(readxl::read_excel, path = path) %>%
  mutate(Cluster = str_remove(Cluster, "module_")) %>%
  arrange(feature_type)

#convert to a clue partition
multiomics_combined14_single_partition<-multiomics_combined14_single %>%
  pull(Cluster) %>%
  set_names(multiomics_combined14_single$feature_type) %>%
  as.cl_partition()

#convert consensus clustering to a clue hard partition
multiomics_14_partition <- multiomics_14_data_ann %>%
   arrange(feature_type) %>%
  pull(Cluster) %>%
  set_names(multiomics_14_data_ann$feature_type) %>%
  as.cl_partition()

single_to_consensus_agreement <- cl_agreement(multiomics_combined14_single_partition, multiomics_14_partition, method = "FM")

#calculate means for single clustering
multiomics_14_single_features_data_mean<- read_csv(paste0(data_path, "MDD_combined14_single_mean.csv"))
multiomics_14_single_features_data_mean_dm <- multiomics_14_single_features_data_mean %>%
  select(-Cluster) %>%
  as.matrix()
rownames(multiomics_14_single_features_data_mean_dm) <- multiomics_14_single_features_data_mean$Cluster
single_to_consensus_cors <- cor(t(multiomics_14_data_mean_dm), t(multiomics_14_single_features_data_mean_dm), method = "pearson")

hm_single_to_consensus_cors <- Heatmap(single_to_consensus_cors, 
                                            column_title = "single to consensus clustering correlations",
                                            name = "pearson\ncorrelations ",
                                            row_names_gp = gpar(fontsize = 5),
                                            column_names_gp = gpar(fontsize = 5),
                                            col = colorRamp2(c(-1, 0, 1), c("#2166AC", "white", "#B2182B")))
hm_single_to_consensus_cors

#find set differences between single seed modules 2 and 9 and consensus modules 9 and 10 respectively.
cluster2_single <- multiomics_combined14_single %>%
  filter(Cluster == "2")
cluster4_consensus <- multiomics_14_data_ann %>%
  filter(Cluster == "4")
cluster2_inters <- intersect(cluster2_single$feature_type, cluster4_consensus$feature_type)
cluster2_dif <- setdiff(cluster2_single$feature_type, cluster4_consensus$feature_type)
cluster4_dif <- setdiff(cluster4_consensus$feature_type,cluster2_single$feature_type)

multiomics_combined14_single_2_9 <- multiomics_combined14_single %>%
  filter(Cluster %in% c("2","9")) %>%
  rename(Single_Clustering = Cluster) %>%
  select(feature_type, Single_Clustering)
multiomics_14_data_ann_4_10 <- multiomics_14_data_ann %>%
  filter(Cluster %in% c("4", "10")) %>%
  rename(Consensus_Clustering =Cluster) %>%
  mutate(Consensus_Clustering = as.character(Consensus_Clustering)) %>%
select(Consensus_Clustering, feature_type)
clustering_module_comparisons <- multiomics_combined14_single_2_9 %>%
 full_join(multiomics_14_data_ann_4_10, by = "feature_type") %>%
  mutate(Both = case_when(!is.na(Single_Clustering)&!is.na(Consensus_Clustering) ~ "x",
                          TRUE ~ ""))
 write_csv(clustering_module_comparisons, "../multiomics_analysis/tables/MDD_multiomics_cl1_cl9_comparisons.csv")

 clustering_module_comparisons_counts <- clustering_module_comparisons %>%
   group_by(Single_Clustering, Consensus_Clustering, Both) %>%
   count()
 write_csv(clustering_module_comparisons_counts, "../multiomics_analysis/tables/MDD_multiomics_cl1_cl9_comparisons_counts.csv")


```

```{r writeTables}
selectFeature <- function(df){
  res <- df %>%
    select(feature, Type) %>%
    mutate(analyte = str_remove(feature, "_.*")) %>%
    rename(assay = Type) %>%
    select(feature, analyte, assay, time, symbol)
return(res)
}

selectDerivedFeature <- function(df){
  res <- df %>%
    select(feature, FDR, Type)
return(res)
}

multiomics_14 <- multiomics_14_data_ann %>%
  arrange(Cluster, Type) %>%
  mutate(Cluster = paste0("module_", Cluster),
         Cluster = factor(Cluster, levels = unique(Cluster))) %>%
  split(.$Cluster) 

res <- write_xlsx(multiomics_14,
                  path = "../multiomics_analysis/tables/MDD_multiomics_14.xlsx")

res <- multiomics_14_data_ann %>%
  arrange(Cluster, Type) %>%
  mutate(Cluster = paste0("module_", Cluster),
         Cluster = factor(Cluster, levels = unique(Cluster))) %>%
  write_csv(file = "../multiomics_analysis/tables/MDD_multiomics_14.csv")

#res <- write_csv(multiomics_14_data_mean, "../multiomics_analysis/tables/MDD_multiomics_14_mean.csv")

```

#
### phenotype analysis

```{r correlate_to_phenotypes}

#start with level 3 IF data from synapse
if_df <- read_csv("../Immunofluorescence/Data/MDD_IF_Level3.csv",skip = 1) %>%
  rename("feature" = "specimenName") %>%
  filter(str_detect(feature, "_")) %>%
  select(matches("feature|24_C2|48_C2")) %>%
  mutate(across(-feature, as.numeric)) %>%
  pivot_longer(cols = matches("_C2_")) %>%
  mutate(condition = str_remove(name, "_C2_.$")) %>%
  mutate(condition = str_replace(condition, "BMP2","BMP2+EGF"),
         condition = str_replace(condition, "IFNG","IFNG+EGF"),
         condition = str_replace(condition, "TGFB","TGFB+EGF")) %>%
  group_by(feature, condition) %>%
  summarise(value = mean(value), .groups = "drop") %>%
  ungroup() %>%
  pivot_wider(names_from = condition, values_from = value) %>%
  select(feature, all_of(condition_order))

if_df_migration <- read_csv("../Immunofluorescence/Data//MDD_migration_distance_lineage_T0.csv") %>%
  filter(thour %in%c(24, 48),
         !treatment == "PBS") %>%
  mutate(condition = paste0(treatment,"_",thour),
         feature = "migration_distance")  %>%
  select(-treatment, -thour) %>%
  pivot_wider(names_from = condition, values_from = lineage_length_mean) %>%
  bind_rows(if_df) 

phenotype_data_ann <- if_df_migration %>%
  filter(str_detect(feature, "Well_Cell_Count|Mean_Cells_per_Cluster|Normalized_Second_Neighbor_Dist|DNA2n_Proportion|Edu_Positive_Proportion|MeanIntensity_KRT5|migration_distance")) %>%
  pivot_longer(cols = -feature, names_to = "condition", values_to = "phenotype_value") %>%
  group_by(feature) %>%
  mutate(phenotype_value = scale(phenotype_value)) %>%
  ungroup()

module_phenotype_data_ann <- multiomics_14_data_mean %>%
  pivot_longer(cols = -Cluster, names_to = "condition", values_to = "module_value") %>%
  mutate(ligand = str_remove(condition, "_.*"),
         time = str_remove(condition, ".*_") %>% as.integer) %>%
  rename(module = Cluster) %>%
  inner_join(phenotype_data_ann, by = "condition")

df <- module_phenotype_data_ann %>%
  filter(module == "10",
         feature == "Edu_Positive_Proportion")

p_mod10_edu <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm", formula = 'y ~x') +
  geom_point(aes(color = ligand, size = time)) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 10 EdU positive proportion vs module",
       x = "Module expression",
       y = "EdU positive proportion") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod10_edu

df <- module_phenotype_data_ann %>%
  filter(module == "9+18",
         feature == "Mean_Cells_per_Cluster")

p_mod9_18_mcc <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm", formula = 'y ~x') +
  geom_point(aes(color = ligand, size = time)) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 9+18 Mean_Cells_per_Cluster proportion vs module",
       x = "Module expression",
       y = "Mean cells per cluster") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod9_18_mcc

df <- module_phenotype_data_ann %>%
  filter(module == "7",
         feature == "Mean_Cells_per_Cluster")

p_mod7_mcc <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm", formula = 'y ~x') +
  geom_point(aes(color = ligand, size = time)) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 7 Mean_Cells_per_Cluster proportion vs module",
       x = "Module expression",
       y = "Mean cells per cluster") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod7_mcc

df <- module_phenotype_data_ann %>%
  filter(module == "2+8",
         feature == "Normalized_Second_Neighbor_Dist")

p_mod2_8_nsnd <- ggplot(df, aes(x = module_value, y = phenotype_value)) +
  stat_smooth(se = FALSE, method = "lm", formula = 'y ~x') +
  geom_point(aes(color = ligand, size = time)) +
  scale_color_manual(values = ligand_cols) +
  scale_radius(breaks = c(0,1,4, 8, 24, 48), range = c(4, 6)) +
  labs(title = "Module 2+8 Normalized_Second_Neighbor_Dist proportion vs module",
       x = "Module expression",
       y = "Normalized second neighbor dist") +
  guides(color = guide_legend(override.aes = list(size = 4)),
         size = guide_legend(override.aes = list(shape = 19))) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text.x = element_text(size = 3.5))
p_mod2_8_nsnd

if(create_pdfs) {
  pdf("../multiomics_analysis/plots/MDD_F5d.pdf", useDingbats = TRUE)
  res <- print(p_mod10_edu)
  res <- dev.off()
  pdf("../multiomics_analysis/plots/MDD_F5e.pdf", useDingbats = TRUE)
  res <- print(p_mod9_18_mcc)
  res <- dev.off() 
  pdf("../multiomics_analysis/plots/MDD_F5f.pdf", useDingbats = TRUE)
  res <- print(p_mod7_mcc)
  res <- dev.off() 
  pdf("../multiomics_analysis/plots/MDD_F5g.pdf", useDingbats = TRUE)
  res <- print(p_mod2_8_nsnd)
  res <- dev.off()
}

```


### cluster analysis


```{r featureTypes, fig.width=8, fig.height=4}

#create a bar chart of the multiomics_14 feature types
df <- multiomics_14_data_ann %>%
  select(feature, Type) %>%
  distinct()

p_log <- ggplot(df, aes(x = Type, fill = Type)) +
  geom_bar() +
  scale_y_log10() +
  scale_fill_manual(values = assay_cols)+
  labs(title = "Multiomics assay features",
       x ="",
       y = "Feature Count (log scaled)") +
  theme_bw()+
  theme(axis.text.x =  element_text(angle = 90),
    axis.ticks = element_blank(),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank())

p_log
if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_SF6a.pdf")
  print(p_log)
  res <- dev.off()
} 


```


```{r gapStatistics, fig.width=4, fig.height=4, eval = TRUE}

plot_gap_custom <- function(x) {
  gstab = data.frame(x$Tab, k = seq_len(nrow(x$Tab)))
  p <- ggplot(gstab, aes(k, gap)) + 
    geom_line() +
    geom_errorbar(aes(ymax = gap + SE.sim,
                      ymin = gap - SE.sim), width=0.1) +
    geom_point(size = .5, col=  "red") +
    labs(title = paste("Gap analysis,",cluster_method)) +
     theme_bw()+
  theme(axis.ticks = element_blank(),
        panel.border = element_blank(),
        axis.line =element_blank(),
        title = element_text(size = 8))
  return(p)
}

dm <- multiomics_data %>%
    select(all_of(condition_order)) %>%
  as.matrix()

if(file.exists(paste0("../multiomics_analysis/data/gss_",cluster_method,"_combined18.rda"))){
  load(paste0("../multiomics_analysis/data/gss_",cluster_method,"_combined18.rda"))
} else {
  gss_combined18 <- clusGap(dm, FUN = pam, pamonce = 5, K.max = 25, B = 100, verbose = FALSE)
  save(gss_combined18, file = paste0("../multiomics_analysis/data/gss_",cluster_method,"_combined18.rda"))
}

  plot_gap_custom(gss_combined18)
if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_SF6c.pdf")
  p <- plot_gap_custom(gss_combined18)
  print(p)
  res <- dev.off()
}


```


This analysis can be interpreted to select the number of clusters in the dataset. It is based on a method described in Modern Statistics for Modern Biology by Susan Holmes and Wolfgang Huber http://web.stanford.edu/class/bios221/book/Chap-Clustering.html and is excerpted below.  

Taking the logarithm of the within-sum-of-squares (log(WSSk)) and comparing it to averages from simulated data with less structure can be a good way of choosing k. This is the basic idea of the gap statistic introduced by Tibshirani, Walther, and Hastie (2001). We compute log(WSSk) for a range of values of k, the number of clusters, and compare it to that obtained on reference data of similar dimensions with various possible ‘non-clustered’ distributions. We can use uniformly distributed data as we did above or data simulated with the same covariance structure as our original data.

The default choice for the number of clusters, k1, is the first value of k for which the gap is not larger than the first local maximum minus a standard error (see the manual page of the clusGap function). ... the choice recommended by Tibshirani, Walther, and Hastie (2001) is the smallest k such that gap(k)≥gap(k+1)−s′k+1...

#### combined18 dataset analysis  

Three of the methods for choosing k using a gap statitistic are:  
"Tibs2001SEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("Tibs2001SEmax"),
      SE.factor = 1) `  
      
uses the criterion, Tibshirani et al (2001) proposed: “the smallest k such that f(k) ≥ f(k+1) - s_{k+1}”. Note that this chooses k = 1 when all standard deviations are larger than the differences f(k+1) - f(k).


"firstSEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("firstSEmax"),
      SE.factor = 1) `  
      
location of the first f() value which is not smaller than the first local maximum minus SE.factor * SE.f[], i.e, within an “f S.E.” range of that maximum (see also SE.factor).

This, the default, has been proposed by Martin Maechler in 2012, when adding clusGap() to the cluster package, after having seen the "globalSEmax" proposal (in code) and read the "Tibs2001SEmax" proposal.

"globalSEmax": `r maxSE(gss_combined18$Tab[,'gap'], SE.f = gss_combined18$Tab[,'SE.sim'],
      method = c("globalSEmax"),
      SE.factor = 1)`    
      
(used in Dudoit and Fridlyand (2002), supposedly following Tibshirani's proposition): location of the first f() value which is not smaller than the global maximum minus SE.factor * SE.f[], i.e, within an “f S.E.” range of that maximum (see also SE.factor).




```{r assayClusterdistribution, fig.width=8, fig.height=4, eval = TRUE}

#multiomics_14 cluster analysis
df <- multiomics_14_data_ann %>%
  mutate(Type = case_when(Type == "motifs" ~"ATAC motifs",
                          TRUE ~ Type))
p_hist <- ggplot(df, aes(x = as.factor(Cluster))) +
  geom_bar(position = "dodge", fill = "black", width = .8) +  
  #scale_fill_manual(values = assay_cols) +
  labs(title = "Feature distribution across modules",
       x = "",
       y = "Feature count") +
  guides(fill = "none")+
  facet_wrap(~Type, scales = "free_y") +
  theme_bw()+
  theme(#axis.title = element_blank(),
    #axis.text = element_blank(),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines"))
p_hist

if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_SF6d.pdf", width = 5, height = 3)
    print(p_hist)
  res <- dev.off()
} 

# Color bars by treatment using standard ligand colors
ligand_facet_order <- c( "EGF", "HGF", "OSM", "BMP2+EGF", "IFNG+EGF", "TGFB+EGF")
df <- multiomics_14_data_mean %>%
  pivot_longer(cols = contains("_")) %>%
  mutate(ligand = str_remove(name, "_.*"),
         ligand = factor(ligand, levels = ligand_facet_order)) %>%
  group_by(Cluster, ligand) %>%
  summarise(mean = mean(value), .groups = "drop")

p_ligand <- ggplot(df, aes(x = Cluster, y =mean, fill = ligand))+
  geom_col() +
  theme_bw() +
  scale_fill_manual(values = ligand_cols) +
  theme(axis.ticks = element_blank(),
    axis.text.x = element_text(size = 5, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line =element_blank(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines")) +
  facet_wrap(~ligand)
p_ligand

if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_SF6e.pdf", width = 4, height = 3)
  print(p_ligand)
  res <- dev.off()
} 

#Module 2 only
df <- multiomics_14_data_mean %>%
  pivot_longer(cols = contains("_")) %>%
  filter(Cluster == 4) %>%
  mutate(ligand = str_remove(name, "_.*")) %>%
  group_by(Cluster, ligand) %>%
  summarise(mean = mean(value), .groups = "drop")

cl4_mean <- ggplot(df, aes(x = ligand, y =mean, fill = ligand))+
  geom_col() +
  theme_bw() +
  labs(x = "",
       y = "Mean Module 4 Expression") +
  guides(fill = "none") +
  scale_fill_manual(values = ligand_cols) +
  scale_y_continuous(breaks = c(-1.5, -1, -.5, 0, .5, 1, 1.5),
                     labels = c("", -1, "", 0, "", 1, "")) +
  theme(axis.ticks.x = element_blank(),
    axis.text.x = element_text(size = 10, angle = 90, color = "black"),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x =element_blank(),
    axis.line.y = element_line(),
    axis.ticks.y = element_line(),
    strip.background = element_blank(),
    legend.key.size = unit(.5, "lines"))
cl4_mean

if(create_pdfs){
  pdf("../multiomics_analysis/plots/MDD_F7b.pdf", width = 4, height = 3)
  print(cl4_mean)
  res <- dev.off()
} 

```

